// sst@2.24.28/constructs/Script.js
import path from "path";
import url from "url";
import { Construct } from "constructs";
import { CustomResource, Duration } from "aws-cdk-lib/core";
import { PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Code, Runtime, Function as CdkFunction } from "aws-cdk-lib/aws-lambda";
import { Stack } from "./Stack.js";
import { Function as Fn, } from "./Function.js";
import { getFunctionRef, } from "./Construct.js";
const __dirname = path.dirname(url.fileURLToPath(import.meta.url));
/////////////////////
// Construct
/////////////////////
/**
 * The `Script` construct is a higher level CDK construct that makes it easy to run a script in a Lambda function during the deployment process.
 *
 * @example
 *
 * ```js
 * import { Script } from "sst/constructs";
 *
 * new Script(stack, "Script", {
 *   onCreate: "src/function.create",
 *   onUpdate: "src/function.update",
 *   onDelete: "src/function.delete",
 * });
 * ```
 */
export class Script extends Construct {
    /**
     * The internally created onCreate `Function` instance.
     */
    createFunction;
    /**
     * The internally created onUpdate `Function` instance.
     */
    updateFunction;
    /**
     * The internally created onDelete `Function` instance.
     */
    deleteFunction;
    props;
    id;
    constructor(scope, id, props) {
        super(scope, id);
        this.id = id;
        if (props.function)
            this.checkDeprecatedFunction();
        // Validate deprecated "function" prop
        // Validate at least 1 function is provided
        if (!props.onCreate && !props.onUpdate && !props.onDelete) {
            throw new Error(`Need to provide at least one of "onCreate", "onUpdate", or "onDelete" functions for the "${this.node.id}" Script`);
        }
        const root = scope.node.root;
        this.props = props;
        this.createFunction = this.createUserFunction("onCreate", props.onCreate);
        this.updateFunction = this.createUserFunction("onUpdate", props.onUpdate);
        this.deleteFunction = this.createUserFunction("onDelete", props.onDelete);
        const crFunction = this.createCustomResourceFunction();
        this.createCustomResource(root, crFunction);
    }
    /**
     * Binds additional resources to the script
     *
     * @example
     * ```js
     * script.bind([STRIPE_KEY, bucket]);
     * ```
     */
    bind(constructs) {
        this.createFunction?.bind(constructs);
        this.updateFunction?.bind(constructs);
        this.deleteFunction?.bind(constructs);
    }
    /**
     * Grants additional permissions to the script
     *
     * @example
     * ```js
     * script.attachPermissions(["s3"]);
     * ```
     */
    attachPermissions(permissions) {
        this.createFunction?.attachPermissions(permissions);
        this.updateFunction?.attachPermissions(permissions);
        this.deleteFunction?.attachPermissions(permissions);
    }
    createUserFunction(type, fnDef) {
        if (!fnDef) {
            return;
        }
        // function is construct => return function directly
        if (fnDef instanceof Fn) {
            // validate live dev is not enabled
            if (fnDef._isLiveDevEnabled) {
                throw new Error(`Live Lambda Dev cannot be enabled for functions in the Script construct. Set the "enableLiveDev" prop for the function to "false".`);
            }
            return Fn.fromDefinition(this, `${type}Function`, fnDef, this.props.defaults?.function, `The "defaults.function" cannot be applied if an instance of a Function construct is passed in. Make sure to define the "${type}" function using FunctionProps, so the Script construct can apply the "defaults.function" to them.`);
        }
        // function is string => create function
        else if (typeof fnDef === "string") {
            return Fn.fromDefinition(this, `${type}Function`, {
                handler: fnDef,
                enableLiveDev: false,
            }, {
                timeout: 900,
                ...this.props.defaults?.function,
            });
        }
        // function is props => create function
        return Fn.fromDefinition(this, `${type}Function`, {
            ...fnDef,
            enableLiveDev: false,
        }, {
            timeout: 900,
            ...this.props.defaults?.function,
        });
    }
    createCustomResourceFunction() {
        const handler = new CdkFunction(this, "ScriptHandler", {
            code: Code.fromAsset(path.join(__dirname, "../support/script-function")),
            runtime: Runtime.NODEJS_16_X,
            handler: "index.handler",
            timeout: Duration.minutes(15),
            memorySize: 1024,
            initialPolicy: [
                new PolicyStatement({
                    actions: ["cloudformation:DescribeStacks"],
                    resources: [Stack.of(this).stackId],
                }),
            ],
        });
        this.createFunction?.grantInvoke(handler);
        this.updateFunction?.grantInvoke(handler);
        this.deleteFunction?.grantInvoke(handler);
        return handler;
    }
    createCustomResource(app, crFunction) {
        // Note: "Version" is set to current timestamp to ensure the Custom
        //       Resource function is run on every update.
        //
        //       Do not use the current timestamp in Live mode, b/c we want the
        //       this custom resource to remain the same in CloudFormation template
        //       when rebuilding infrastructure. Otherwise, there will always be
        //       a change when rebuilding infrastructure b/c the "version" property
        //       changes on each build.
        const defaultVersion = app.mode === "dev" ? app.debugScriptVersion : Date.now().toString();
        const version = this.props.version ?? defaultVersion;
        new CustomResource(this, "ScriptResource", {
            serviceToken: crFunction.functionArn,
            resourceType: "Custom::SSTScript",
            properties: {
                UserCreateFunction: this.createFunction?.functionName,
                UserUpdateFunction: this.updateFunction?.functionName,
                UserDeleteFunction: this.deleteFunction?.functionName,
                UserParams: JSON.stringify(this.props.params || {}),
                Version: version,
            },
        });
    }
    checkDeprecatedFunction() {
        throw new Error(`The "function" property has been replaced by "onCreate" and "onUpdate". More details on upgrading - https://docs.sst.dev/constructs/Script#upgrading-to-v0460`);
    }
    /** @internal */
    getConstructMetadata() {
        return {
            type: "Script",
            data: {
                createfn: getFunctionRef(this.createFunction),
                deletefn: getFunctionRef(this.deleteFunction),
                updatefn: getFunctionRef(this.updateFunction),
            },
        };
    }
    /** @internal */
    getFunctionBinding() {
        return undefined;
    }
}
