// sst@2.36.1/constructs/future/Auth.js
import { Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Construct } from "constructs";
import { Api } from "../Api.js";
import { Stack } from "../Stack.js";
import { Secret } from "../Secret.js";
import { getParameterPath, } from "../util/functionBinding.js";
import { CustomResource } from "aws-cdk-lib/core";
/**
 * SST Auth is a lightweight authentication solution for your applications. With a simple set of configuration you can deploy a function attached to your API that can handle various authentication flows.  *
 * @example
 * ```
 * import { Auth } from "sst/constructs"
 *
 * new Auth(stack, "auth", {
 *   authenticator: "functions/authenticator.handler"
 * })
 */
export class Auth extends Construct {
    id;
    authenticator;
    api;
    publicKey;
    privateKey;
    constructor(scope, id, props) {
        super(scope, props.cdk?.id || id);
        if (!props.authenticator ||
            "defaults" in props ||
            "login" in props ||
            "triggers" in props ||
            "identityPoolFederation" in props ||
            "cdk" in props) {
            throw new Error(`It looks like you may be passing in Cognito props to the Auth construct. The Auth construct was renamed to Cognito in version 1.10.0`);
        }
        this.id = id;
        const stack = Stack.of(scope);
        this.authenticator = props.authenticator;
        this.api = new Api(this, id + "-api", {
            defaults: {
                function: {
                    environment: {
                        AUTH_ID: id,
                    },
                },
            },
            routes: {
                "ANY /{step}": {
                    function: this.authenticator,
                },
                "ANY /": {
                    function: this.authenticator,
                },
            },
            customDomain: props.customDomain,
        });
        this.publicKey = new Secret(this, id + "PublicKey");
        this.privateKey = new Secret(this, id + "PrivateKey");
        const fn = this.api.getFunction("ANY /{step}");
        fn.bind([this.publicKey, this.privateKey]);
        const app = this.node.root;
        fn.attachPermissions([
            new PolicyStatement({
                actions: ["ssm:GetParameters"],
                effect: Effect.ALLOW,
                resources: [
                    `arn:${Stack.of(this).partition}:ssm:${app.region}:${app.account}:parameter${getParameterPath(this.publicKey, "value")}`,
                    `arn:${Stack.of(this).partition}:ssm:${app.region}:${app.account}:parameter${getParameterPath(this.privateKey, "value")}`,
                ],
            }),
        ]);
        const policy = new Policy(this, "CloudFrontInvalidatorPolicy", {
            statements: [
                new PolicyStatement({
                    actions: [
                        "ssm:GetParameter",
                        "ssm:PutParameter",
                        "ssm:DeleteParameter",
                        "ssm:DeleteParameters",
                    ],
                    resources: [
                        `arn:${stack.partition}:ssm:${stack.region}:${stack.account}:parameter/*`,
                    ],
                }),
            ],
        });
        stack.customResourceHandler.role?.attachInlinePolicy(policy);
        const resource = new CustomResource(this, "StackMetadata", {
            serviceToken: stack.customResourceHandler.functionArn,
            resourceType: "Custom::AuthKeys",
            properties: {
                publicPath: getParameterPath(this.publicKey, "value"),
                privatePath: getParameterPath(this.privateKey, "value"),
            },
        });
        resource.node.addDependency(policy);
        app.registerTypes(this);
    }
    get url() {
        return this.api.customDomainUrl || this.api.url;
    }
    /** @internal */
    getConstructMetadata() {
        return {
            type: "Auth",
            data: {},
        };
    }
    /** @internal */
    getFunctionBinding() {
        return {
            clientPackage: "future/auth",
            variables: {
                publicKey: {
                    type: "secret_reference",
                    secret: this.publicKey,
                },
                url: {
                    type: "plain",
                    value: this.url,
                },
            },
            permissions: {},
        };
    }
}
