// kysely-codegen@0.10.1/dist/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Generator = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const perf_hooks_1 = require("perf_hooks");
const serializer_1 = require("./serializer");
const transformer_1 = require("./transformer");
/**
 * Generates codegen output using specified options.
 */
class Generator {
    async generate(options) {
        const startTime = perf_hooks_1.performance.now();
        options.logger?.info('Introspecting database...');
        const metadata = await options.dialect.introspector.introspect({
            db: options.db,
            excludePattern: options.excludePattern,
            includePattern: options.includePattern,
        });
        options.logger?.debug();
        options.logger?.debug(`Found ${metadata.tables.length} public tables:`);
        for (const table of metadata.tables) {
            options.logger?.debug(` - ${table.name}`);
        }
        options.logger?.debug();
        const transformer = options.transformer ?? new transformer_1.Transformer();
        const nodes = transformer.transform({
            camelCase: !!options.camelCase,
            defaultSchema: options.schema,
            dialect: options.dialect,
            metadata,
        });
        const serializer = options.serializer ??
            new serializer_1.Serializer({ typeOnlyImports: options.typeOnlyImports });
        const data = serializer.serialize(nodes);
        if (options.print) {
            console.log();
            console.log(data);
        }
        else if (options.outFile) {
            const outDir = (0, path_1.parse)(options.outFile).dir;
            await fs_1.promises.mkdir(outDir, { recursive: true });
            await fs_1.promises.writeFile(options.outFile, data);
            const endTime = perf_hooks_1.performance.now();
            const relativeOutDir = `.${path_1.sep}${(0, path_1.relative)(process.cwd(), options.outFile)}`;
            const duration = Math.round(endTime - startTime);
            const tableCount = metadata.tables.length;
            const s = tableCount === 1 ? '' : 's';
            options.logger?.success(`Introspected ${tableCount} table${s} and generated ${relativeOutDir} in ${duration}ms.\n`);
        }
        return data;
    }
}
exports.Generator = Generator;
