// @aws-sdk/client-iot-data-plane@3.449.0/dist-cjs/models/models_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestEntityTooLargeException = exports.PayloadFormatIndicator = exports.UnsupportedDocumentEncodingException = exports.UnauthorizedException = exports.ThrottlingException = exports.ServiceUnavailableException = exports.ResourceNotFoundException = exports.MethodNotAllowedException = exports.InvalidRequestException = exports.InternalFailureException = exports.ConflictException = void 0;
const IoTDataPlaneServiceException_1 = require("./IoTDataPlaneServiceException");
class ConflictException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "ConflictException",
            $fault: "client",
            ...opts,
        });
        this.name = "ConflictException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ConflictException.prototype);
    }
}
exports.ConflictException = ConflictException;
class InternalFailureException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "InternalFailureException",
            $fault: "server",
            ...opts,
        });
        this.name = "InternalFailureException";
        this.$fault = "server";
        Object.setPrototypeOf(this, InternalFailureException.prototype);
    }
}
exports.InternalFailureException = InternalFailureException;
class InvalidRequestException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "InvalidRequestException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRequestException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRequestException.prototype);
    }
}
exports.InvalidRequestException = InvalidRequestException;
class MethodNotAllowedException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "MethodNotAllowedException",
            $fault: "client",
            ...opts,
        });
        this.name = "MethodNotAllowedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MethodNotAllowedException.prototype);
    }
}
exports.MethodNotAllowedException = MethodNotAllowedException;
class ResourceNotFoundException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceNotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
exports.ResourceNotFoundException = ResourceNotFoundException;
class ServiceUnavailableException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "ServiceUnavailableException",
            $fault: "server",
            ...opts,
        });
        this.name = "ServiceUnavailableException";
        this.$fault = "server";
        Object.setPrototypeOf(this, ServiceUnavailableException.prototype);
    }
}
exports.ServiceUnavailableException = ServiceUnavailableException;
class ThrottlingException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        this.name = "ThrottlingException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ThrottlingException.prototype);
    }
}
exports.ThrottlingException = ThrottlingException;
class UnauthorizedException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "UnauthorizedException",
            $fault: "client",
            ...opts,
        });
        this.name = "UnauthorizedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, UnauthorizedException.prototype);
    }
}
exports.UnauthorizedException = UnauthorizedException;
class UnsupportedDocumentEncodingException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "UnsupportedDocumentEncodingException",
            $fault: "client",
            ...opts,
        });
        this.name = "UnsupportedDocumentEncodingException";
        this.$fault = "client";
        Object.setPrototypeOf(this, UnsupportedDocumentEncodingException.prototype);
    }
}
exports.UnsupportedDocumentEncodingException = UnsupportedDocumentEncodingException;
exports.PayloadFormatIndicator = {
    UNSPECIFIED_BYTES: "UNSPECIFIED_BYTES",
    UTF8_DATA: "UTF8_DATA",
};
class RequestEntityTooLargeException extends IoTDataPlaneServiceException_1.IoTDataPlaneServiceException {
    constructor(opts) {
        super({
            name: "RequestEntityTooLargeException",
            $fault: "client",
            ...opts,
        });
        this.name = "RequestEntityTooLargeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RequestEntityTooLargeException.prototype);
    }
}
exports.RequestEntityTooLargeException = RequestEntityTooLargeException;
