// @aws-sdk/client-iot@3.449.0/dist-cjs/IoT.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoT = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AcceptCertificateTransferCommand_1 = require("./commands/AcceptCertificateTransferCommand");
const AddThingToBillingGroupCommand_1 = require("./commands/AddThingToBillingGroupCommand");
const AddThingToThingGroupCommand_1 = require("./commands/AddThingToThingGroupCommand");
const AssociateTargetsWithJobCommand_1 = require("./commands/AssociateTargetsWithJobCommand");
const AttachPolicyCommand_1 = require("./commands/AttachPolicyCommand");
const AttachPrincipalPolicyCommand_1 = require("./commands/AttachPrincipalPolicyCommand");
const AttachSecurityProfileCommand_1 = require("./commands/AttachSecurityProfileCommand");
const AttachThingPrincipalCommand_1 = require("./commands/AttachThingPrincipalCommand");
const CancelAuditMitigationActionsTaskCommand_1 = require("./commands/CancelAuditMitigationActionsTaskCommand");
const CancelAuditTaskCommand_1 = require("./commands/CancelAuditTaskCommand");
const CancelCertificateTransferCommand_1 = require("./commands/CancelCertificateTransferCommand");
const CancelDetectMitigationActionsTaskCommand_1 = require("./commands/CancelDetectMitigationActionsTaskCommand");
const CancelJobCommand_1 = require("./commands/CancelJobCommand");
const CancelJobExecutionCommand_1 = require("./commands/CancelJobExecutionCommand");
const ClearDefaultAuthorizerCommand_1 = require("./commands/ClearDefaultAuthorizerCommand");
const ConfirmTopicRuleDestinationCommand_1 = require("./commands/ConfirmTopicRuleDestinationCommand");
const CreateAuditSuppressionCommand_1 = require("./commands/CreateAuditSuppressionCommand");
const CreateAuthorizerCommand_1 = require("./commands/CreateAuthorizerCommand");
const CreateBillingGroupCommand_1 = require("./commands/CreateBillingGroupCommand");
const CreateCertificateFromCsrCommand_1 = require("./commands/CreateCertificateFromCsrCommand");
const CreateCustomMetricCommand_1 = require("./commands/CreateCustomMetricCommand");
const CreateDimensionCommand_1 = require("./commands/CreateDimensionCommand");
const CreateDomainConfigurationCommand_1 = require("./commands/CreateDomainConfigurationCommand");
const CreateDynamicThingGroupCommand_1 = require("./commands/CreateDynamicThingGroupCommand");
const CreateFleetMetricCommand_1 = require("./commands/CreateFleetMetricCommand");
const CreateJobCommand_1 = require("./commands/CreateJobCommand");
const CreateJobTemplateCommand_1 = require("./commands/CreateJobTemplateCommand");
const CreateKeysAndCertificateCommand_1 = require("./commands/CreateKeysAndCertificateCommand");
const CreateMitigationActionCommand_1 = require("./commands/CreateMitigationActionCommand");
const CreateOTAUpdateCommand_1 = require("./commands/CreateOTAUpdateCommand");
const CreatePackageCommand_1 = require("./commands/CreatePackageCommand");
const CreatePackageVersionCommand_1 = require("./commands/CreatePackageVersionCommand");
const CreatePolicyCommand_1 = require("./commands/CreatePolicyCommand");
const CreatePolicyVersionCommand_1 = require("./commands/CreatePolicyVersionCommand");
const CreateProvisioningClaimCommand_1 = require("./commands/CreateProvisioningClaimCommand");
const CreateProvisioningTemplateCommand_1 = require("./commands/CreateProvisioningTemplateCommand");
const CreateProvisioningTemplateVersionCommand_1 = require("./commands/CreateProvisioningTemplateVersionCommand");
const CreateRoleAliasCommand_1 = require("./commands/CreateRoleAliasCommand");
const CreateScheduledAuditCommand_1 = require("./commands/CreateScheduledAuditCommand");
const CreateSecurityProfileCommand_1 = require("./commands/CreateSecurityProfileCommand");
const CreateStreamCommand_1 = require("./commands/CreateStreamCommand");
const CreateThingCommand_1 = require("./commands/CreateThingCommand");
const CreateThingGroupCommand_1 = require("./commands/CreateThingGroupCommand");
const CreateThingTypeCommand_1 = require("./commands/CreateThingTypeCommand");
const CreateTopicRuleCommand_1 = require("./commands/CreateTopicRuleCommand");
const CreateTopicRuleDestinationCommand_1 = require("./commands/CreateTopicRuleDestinationCommand");
const DeleteAccountAuditConfigurationCommand_1 = require("./commands/DeleteAccountAuditConfigurationCommand");
const DeleteAuditSuppressionCommand_1 = require("./commands/DeleteAuditSuppressionCommand");
const DeleteAuthorizerCommand_1 = require("./commands/DeleteAuthorizerCommand");
const DeleteBillingGroupCommand_1 = require("./commands/DeleteBillingGroupCommand");
const DeleteCACertificateCommand_1 = require("./commands/DeleteCACertificateCommand");
const DeleteCertificateCommand_1 = require("./commands/DeleteCertificateCommand");
const DeleteCustomMetricCommand_1 = require("./commands/DeleteCustomMetricCommand");
const DeleteDimensionCommand_1 = require("./commands/DeleteDimensionCommand");
const DeleteDomainConfigurationCommand_1 = require("./commands/DeleteDomainConfigurationCommand");
const DeleteDynamicThingGroupCommand_1 = require("./commands/DeleteDynamicThingGroupCommand");
const DeleteFleetMetricCommand_1 = require("./commands/DeleteFleetMetricCommand");
const DeleteJobCommand_1 = require("./commands/DeleteJobCommand");
const DeleteJobExecutionCommand_1 = require("./commands/DeleteJobExecutionCommand");
const DeleteJobTemplateCommand_1 = require("./commands/DeleteJobTemplateCommand");
const DeleteMitigationActionCommand_1 = require("./commands/DeleteMitigationActionCommand");
const DeleteOTAUpdateCommand_1 = require("./commands/DeleteOTAUpdateCommand");
const DeletePackageCommand_1 = require("./commands/DeletePackageCommand");
const DeletePackageVersionCommand_1 = require("./commands/DeletePackageVersionCommand");
const DeletePolicyCommand_1 = require("./commands/DeletePolicyCommand");
const DeletePolicyVersionCommand_1 = require("./commands/DeletePolicyVersionCommand");
const DeleteProvisioningTemplateCommand_1 = require("./commands/DeleteProvisioningTemplateCommand");
const DeleteProvisioningTemplateVersionCommand_1 = require("./commands/DeleteProvisioningTemplateVersionCommand");
const DeleteRegistrationCodeCommand_1 = require("./commands/DeleteRegistrationCodeCommand");
const DeleteRoleAliasCommand_1 = require("./commands/DeleteRoleAliasCommand");
const DeleteScheduledAuditCommand_1 = require("./commands/DeleteScheduledAuditCommand");
const DeleteSecurityProfileCommand_1 = require("./commands/DeleteSecurityProfileCommand");
const DeleteStreamCommand_1 = require("./commands/DeleteStreamCommand");
const DeleteThingCommand_1 = require("./commands/DeleteThingCommand");
const DeleteThingGroupCommand_1 = require("./commands/DeleteThingGroupCommand");
const DeleteThingTypeCommand_1 = require("./commands/DeleteThingTypeCommand");
const DeleteTopicRuleCommand_1 = require("./commands/DeleteTopicRuleCommand");
const DeleteTopicRuleDestinationCommand_1 = require("./commands/DeleteTopicRuleDestinationCommand");
const DeleteV2LoggingLevelCommand_1 = require("./commands/DeleteV2LoggingLevelCommand");
const DeprecateThingTypeCommand_1 = require("./commands/DeprecateThingTypeCommand");
const DescribeAccountAuditConfigurationCommand_1 = require("./commands/DescribeAccountAuditConfigurationCommand");
const DescribeAuditFindingCommand_1 = require("./commands/DescribeAuditFindingCommand");
const DescribeAuditMitigationActionsTaskCommand_1 = require("./commands/DescribeAuditMitigationActionsTaskCommand");
const DescribeAuditSuppressionCommand_1 = require("./commands/DescribeAuditSuppressionCommand");
const DescribeAuditTaskCommand_1 = require("./commands/DescribeAuditTaskCommand");
const DescribeAuthorizerCommand_1 = require("./commands/DescribeAuthorizerCommand");
const DescribeBillingGroupCommand_1 = require("./commands/DescribeBillingGroupCommand");
const DescribeCACertificateCommand_1 = require("./commands/DescribeCACertificateCommand");
const DescribeCertificateCommand_1 = require("./commands/DescribeCertificateCommand");
const DescribeCustomMetricCommand_1 = require("./commands/DescribeCustomMetricCommand");
const DescribeDefaultAuthorizerCommand_1 = require("./commands/DescribeDefaultAuthorizerCommand");
const DescribeDetectMitigationActionsTaskCommand_1 = require("./commands/DescribeDetectMitigationActionsTaskCommand");
const DescribeDimensionCommand_1 = require("./commands/DescribeDimensionCommand");
const DescribeDomainConfigurationCommand_1 = require("./commands/DescribeDomainConfigurationCommand");
const DescribeEndpointCommand_1 = require("./commands/DescribeEndpointCommand");
const DescribeEventConfigurationsCommand_1 = require("./commands/DescribeEventConfigurationsCommand");
const DescribeFleetMetricCommand_1 = require("./commands/DescribeFleetMetricCommand");
const DescribeIndexCommand_1 = require("./commands/DescribeIndexCommand");
const DescribeJobCommand_1 = require("./commands/DescribeJobCommand");
const DescribeJobExecutionCommand_1 = require("./commands/DescribeJobExecutionCommand");
const DescribeJobTemplateCommand_1 = require("./commands/DescribeJobTemplateCommand");
const DescribeManagedJobTemplateCommand_1 = require("./commands/DescribeManagedJobTemplateCommand");
const DescribeMitigationActionCommand_1 = require("./commands/DescribeMitigationActionCommand");
const DescribeProvisioningTemplateCommand_1 = require("./commands/DescribeProvisioningTemplateCommand");
const DescribeProvisioningTemplateVersionCommand_1 = require("./commands/DescribeProvisioningTemplateVersionCommand");
const DescribeRoleAliasCommand_1 = require("./commands/DescribeRoleAliasCommand");
const DescribeScheduledAuditCommand_1 = require("./commands/DescribeScheduledAuditCommand");
const DescribeSecurityProfileCommand_1 = require("./commands/DescribeSecurityProfileCommand");
const DescribeStreamCommand_1 = require("./commands/DescribeStreamCommand");
const DescribeThingCommand_1 = require("./commands/DescribeThingCommand");
const DescribeThingGroupCommand_1 = require("./commands/DescribeThingGroupCommand");
const DescribeThingRegistrationTaskCommand_1 = require("./commands/DescribeThingRegistrationTaskCommand");
const DescribeThingTypeCommand_1 = require("./commands/DescribeThingTypeCommand");
const DetachPolicyCommand_1 = require("./commands/DetachPolicyCommand");
const DetachPrincipalPolicyCommand_1 = require("./commands/DetachPrincipalPolicyCommand");
const DetachSecurityProfileCommand_1 = require("./commands/DetachSecurityProfileCommand");
const DetachThingPrincipalCommand_1 = require("./commands/DetachThingPrincipalCommand");
const DisableTopicRuleCommand_1 = require("./commands/DisableTopicRuleCommand");
const EnableTopicRuleCommand_1 = require("./commands/EnableTopicRuleCommand");
const GetBehaviorModelTrainingSummariesCommand_1 = require("./commands/GetBehaviorModelTrainingSummariesCommand");
const GetBucketsAggregationCommand_1 = require("./commands/GetBucketsAggregationCommand");
const GetCardinalityCommand_1 = require("./commands/GetCardinalityCommand");
const GetEffectivePoliciesCommand_1 = require("./commands/GetEffectivePoliciesCommand");
const GetIndexingConfigurationCommand_1 = require("./commands/GetIndexingConfigurationCommand");
const GetJobDocumentCommand_1 = require("./commands/GetJobDocumentCommand");
const GetLoggingOptionsCommand_1 = require("./commands/GetLoggingOptionsCommand");
const GetOTAUpdateCommand_1 = require("./commands/GetOTAUpdateCommand");
const GetPackageCommand_1 = require("./commands/GetPackageCommand");
const GetPackageConfigurationCommand_1 = require("./commands/GetPackageConfigurationCommand");
const GetPackageVersionCommand_1 = require("./commands/GetPackageVersionCommand");
const GetPercentilesCommand_1 = require("./commands/GetPercentilesCommand");
const GetPolicyCommand_1 = require("./commands/GetPolicyCommand");
const GetPolicyVersionCommand_1 = require("./commands/GetPolicyVersionCommand");
const GetRegistrationCodeCommand_1 = require("./commands/GetRegistrationCodeCommand");
const GetStatisticsCommand_1 = require("./commands/GetStatisticsCommand");
const GetTopicRuleCommand_1 = require("./commands/GetTopicRuleCommand");
const GetTopicRuleDestinationCommand_1 = require("./commands/GetTopicRuleDestinationCommand");
const GetV2LoggingOptionsCommand_1 = require("./commands/GetV2LoggingOptionsCommand");
const ListActiveViolationsCommand_1 = require("./commands/ListActiveViolationsCommand");
const ListAttachedPoliciesCommand_1 = require("./commands/ListAttachedPoliciesCommand");
const ListAuditFindingsCommand_1 = require("./commands/ListAuditFindingsCommand");
const ListAuditMitigationActionsExecutionsCommand_1 = require("./commands/ListAuditMitigationActionsExecutionsCommand");
const ListAuditMitigationActionsTasksCommand_1 = require("./commands/ListAuditMitigationActionsTasksCommand");
const ListAuditSuppressionsCommand_1 = require("./commands/ListAuditSuppressionsCommand");
const ListAuditTasksCommand_1 = require("./commands/ListAuditTasksCommand");
const ListAuthorizersCommand_1 = require("./commands/ListAuthorizersCommand");
const ListBillingGroupsCommand_1 = require("./commands/ListBillingGroupsCommand");
const ListCACertificatesCommand_1 = require("./commands/ListCACertificatesCommand");
const ListCertificatesByCACommand_1 = require("./commands/ListCertificatesByCACommand");
const ListCertificatesCommand_1 = require("./commands/ListCertificatesCommand");
const ListCustomMetricsCommand_1 = require("./commands/ListCustomMetricsCommand");
const ListDetectMitigationActionsExecutionsCommand_1 = require("./commands/ListDetectMitigationActionsExecutionsCommand");
const ListDetectMitigationActionsTasksCommand_1 = require("./commands/ListDetectMitigationActionsTasksCommand");
const ListDimensionsCommand_1 = require("./commands/ListDimensionsCommand");
const ListDomainConfigurationsCommand_1 = require("./commands/ListDomainConfigurationsCommand");
const ListFleetMetricsCommand_1 = require("./commands/ListFleetMetricsCommand");
const ListIndicesCommand_1 = require("./commands/ListIndicesCommand");
const ListJobExecutionsForJobCommand_1 = require("./commands/ListJobExecutionsForJobCommand");
const ListJobExecutionsForThingCommand_1 = require("./commands/ListJobExecutionsForThingCommand");
const ListJobsCommand_1 = require("./commands/ListJobsCommand");
const ListJobTemplatesCommand_1 = require("./commands/ListJobTemplatesCommand");
const ListManagedJobTemplatesCommand_1 = require("./commands/ListManagedJobTemplatesCommand");
const ListMetricValuesCommand_1 = require("./commands/ListMetricValuesCommand");
const ListMitigationActionsCommand_1 = require("./commands/ListMitigationActionsCommand");
const ListOTAUpdatesCommand_1 = require("./commands/ListOTAUpdatesCommand");
const ListOutgoingCertificatesCommand_1 = require("./commands/ListOutgoingCertificatesCommand");
const ListPackagesCommand_1 = require("./commands/ListPackagesCommand");
const ListPackageVersionsCommand_1 = require("./commands/ListPackageVersionsCommand");
const ListPoliciesCommand_1 = require("./commands/ListPoliciesCommand");
const ListPolicyPrincipalsCommand_1 = require("./commands/ListPolicyPrincipalsCommand");
const ListPolicyVersionsCommand_1 = require("./commands/ListPolicyVersionsCommand");
const ListPrincipalPoliciesCommand_1 = require("./commands/ListPrincipalPoliciesCommand");
const ListPrincipalThingsCommand_1 = require("./commands/ListPrincipalThingsCommand");
const ListProvisioningTemplatesCommand_1 = require("./commands/ListProvisioningTemplatesCommand");
const ListProvisioningTemplateVersionsCommand_1 = require("./commands/ListProvisioningTemplateVersionsCommand");
const ListRelatedResourcesForAuditFindingCommand_1 = require("./commands/ListRelatedResourcesForAuditFindingCommand");
const ListRoleAliasesCommand_1 = require("./commands/ListRoleAliasesCommand");
const ListScheduledAuditsCommand_1 = require("./commands/ListScheduledAuditsCommand");
const ListSecurityProfilesCommand_1 = require("./commands/ListSecurityProfilesCommand");
const ListSecurityProfilesForTargetCommand_1 = require("./commands/ListSecurityProfilesForTargetCommand");
const ListStreamsCommand_1 = require("./commands/ListStreamsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const ListTargetsForPolicyCommand_1 = require("./commands/ListTargetsForPolicyCommand");
const ListTargetsForSecurityProfileCommand_1 = require("./commands/ListTargetsForSecurityProfileCommand");
const ListThingGroupsCommand_1 = require("./commands/ListThingGroupsCommand");
const ListThingGroupsForThingCommand_1 = require("./commands/ListThingGroupsForThingCommand");
const ListThingPrincipalsCommand_1 = require("./commands/ListThingPrincipalsCommand");
const ListThingRegistrationTaskReportsCommand_1 = require("./commands/ListThingRegistrationTaskReportsCommand");
const ListThingRegistrationTasksCommand_1 = require("./commands/ListThingRegistrationTasksCommand");
const ListThingsCommand_1 = require("./commands/ListThingsCommand");
const ListThingsInBillingGroupCommand_1 = require("./commands/ListThingsInBillingGroupCommand");
const ListThingsInThingGroupCommand_1 = require("./commands/ListThingsInThingGroupCommand");
const ListThingTypesCommand_1 = require("./commands/ListThingTypesCommand");
const ListTopicRuleDestinationsCommand_1 = require("./commands/ListTopicRuleDestinationsCommand");
const ListTopicRulesCommand_1 = require("./commands/ListTopicRulesCommand");
const ListV2LoggingLevelsCommand_1 = require("./commands/ListV2LoggingLevelsCommand");
const ListViolationEventsCommand_1 = require("./commands/ListViolationEventsCommand");
const PutVerificationStateOnViolationCommand_1 = require("./commands/PutVerificationStateOnViolationCommand");
const RegisterCACertificateCommand_1 = require("./commands/RegisterCACertificateCommand");
const RegisterCertificateCommand_1 = require("./commands/RegisterCertificateCommand");
const RegisterCertificateWithoutCACommand_1 = require("./commands/RegisterCertificateWithoutCACommand");
const RegisterThingCommand_1 = require("./commands/RegisterThingCommand");
const RejectCertificateTransferCommand_1 = require("./commands/RejectCertificateTransferCommand");
const RemoveThingFromBillingGroupCommand_1 = require("./commands/RemoveThingFromBillingGroupCommand");
const RemoveThingFromThingGroupCommand_1 = require("./commands/RemoveThingFromThingGroupCommand");
const ReplaceTopicRuleCommand_1 = require("./commands/ReplaceTopicRuleCommand");
const SearchIndexCommand_1 = require("./commands/SearchIndexCommand");
const SetDefaultAuthorizerCommand_1 = require("./commands/SetDefaultAuthorizerCommand");
const SetDefaultPolicyVersionCommand_1 = require("./commands/SetDefaultPolicyVersionCommand");
const SetLoggingOptionsCommand_1 = require("./commands/SetLoggingOptionsCommand");
const SetV2LoggingLevelCommand_1 = require("./commands/SetV2LoggingLevelCommand");
const SetV2LoggingOptionsCommand_1 = require("./commands/SetV2LoggingOptionsCommand");
const StartAuditMitigationActionsTaskCommand_1 = require("./commands/StartAuditMitigationActionsTaskCommand");
const StartDetectMitigationActionsTaskCommand_1 = require("./commands/StartDetectMitigationActionsTaskCommand");
const StartOnDemandAuditTaskCommand_1 = require("./commands/StartOnDemandAuditTaskCommand");
const StartThingRegistrationTaskCommand_1 = require("./commands/StartThingRegistrationTaskCommand");
const StopThingRegistrationTaskCommand_1 = require("./commands/StopThingRegistrationTaskCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const TestAuthorizationCommand_1 = require("./commands/TestAuthorizationCommand");
const TestInvokeAuthorizerCommand_1 = require("./commands/TestInvokeAuthorizerCommand");
const TransferCertificateCommand_1 = require("./commands/TransferCertificateCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateAccountAuditConfigurationCommand_1 = require("./commands/UpdateAccountAuditConfigurationCommand");
const UpdateAuditSuppressionCommand_1 = require("./commands/UpdateAuditSuppressionCommand");
const UpdateAuthorizerCommand_1 = require("./commands/UpdateAuthorizerCommand");
const UpdateBillingGroupCommand_1 = require("./commands/UpdateBillingGroupCommand");
const UpdateCACertificateCommand_1 = require("./commands/UpdateCACertificateCommand");
const UpdateCertificateCommand_1 = require("./commands/UpdateCertificateCommand");
const UpdateCustomMetricCommand_1 = require("./commands/UpdateCustomMetricCommand");
const UpdateDimensionCommand_1 = require("./commands/UpdateDimensionCommand");
const UpdateDomainConfigurationCommand_1 = require("./commands/UpdateDomainConfigurationCommand");
const UpdateDynamicThingGroupCommand_1 = require("./commands/UpdateDynamicThingGroupCommand");
const UpdateEventConfigurationsCommand_1 = require("./commands/UpdateEventConfigurationsCommand");
const UpdateFleetMetricCommand_1 = require("./commands/UpdateFleetMetricCommand");
const UpdateIndexingConfigurationCommand_1 = require("./commands/UpdateIndexingConfigurationCommand");
const UpdateJobCommand_1 = require("./commands/UpdateJobCommand");
const UpdateMitigationActionCommand_1 = require("./commands/UpdateMitigationActionCommand");
const UpdatePackageCommand_1 = require("./commands/UpdatePackageCommand");
const UpdatePackageConfigurationCommand_1 = require("./commands/UpdatePackageConfigurationCommand");
const UpdatePackageVersionCommand_1 = require("./commands/UpdatePackageVersionCommand");
const UpdateProvisioningTemplateCommand_1 = require("./commands/UpdateProvisioningTemplateCommand");
const UpdateRoleAliasCommand_1 = require("./commands/UpdateRoleAliasCommand");
const UpdateScheduledAuditCommand_1 = require("./commands/UpdateScheduledAuditCommand");
const UpdateSecurityProfileCommand_1 = require("./commands/UpdateSecurityProfileCommand");
const UpdateStreamCommand_1 = require("./commands/UpdateStreamCommand");
const UpdateThingCommand_1 = require("./commands/UpdateThingCommand");
const UpdateThingGroupCommand_1 = require("./commands/UpdateThingGroupCommand");
const UpdateThingGroupsForThingCommand_1 = require("./commands/UpdateThingGroupsForThingCommand");
const UpdateTopicRuleDestinationCommand_1 = require("./commands/UpdateTopicRuleDestinationCommand");
const ValidateSecurityProfileBehaviorsCommand_1 = require("./commands/ValidateSecurityProfileBehaviorsCommand");
const IoTClient_1 = require("./IoTClient");
const commands = {
    AcceptCertificateTransferCommand: AcceptCertificateTransferCommand_1.AcceptCertificateTransferCommand,
    AddThingToBillingGroupCommand: AddThingToBillingGroupCommand_1.AddThingToBillingGroupCommand,
    AddThingToThingGroupCommand: AddThingToThingGroupCommand_1.AddThingToThingGroupCommand,
    AssociateTargetsWithJobCommand: AssociateTargetsWithJobCommand_1.AssociateTargetsWithJobCommand,
    AttachPolicyCommand: AttachPolicyCommand_1.AttachPolicyCommand,
    AttachPrincipalPolicyCommand: AttachPrincipalPolicyCommand_1.AttachPrincipalPolicyCommand,
    AttachSecurityProfileCommand: AttachSecurityProfileCommand_1.AttachSecurityProfileCommand,
    AttachThingPrincipalCommand: AttachThingPrincipalCommand_1.AttachThingPrincipalCommand,
    CancelAuditMitigationActionsTaskCommand: CancelAuditMitigationActionsTaskCommand_1.CancelAuditMitigationActionsTaskCommand,
    CancelAuditTaskCommand: CancelAuditTaskCommand_1.CancelAuditTaskCommand,
    CancelCertificateTransferCommand: CancelCertificateTransferCommand_1.CancelCertificateTransferCommand,
    CancelDetectMitigationActionsTaskCommand: CancelDetectMitigationActionsTaskCommand_1.CancelDetectMitigationActionsTaskCommand,
    CancelJobCommand: CancelJobCommand_1.CancelJobCommand,
    CancelJobExecutionCommand: CancelJobExecutionCommand_1.CancelJobExecutionCommand,
    ClearDefaultAuthorizerCommand: ClearDefaultAuthorizerCommand_1.ClearDefaultAuthorizerCommand,
    ConfirmTopicRuleDestinationCommand: ConfirmTopicRuleDestinationCommand_1.ConfirmTopicRuleDestinationCommand,
    CreateAuditSuppressionCommand: CreateAuditSuppressionCommand_1.CreateAuditSuppressionCommand,
    CreateAuthorizerCommand: CreateAuthorizerCommand_1.CreateAuthorizerCommand,
    CreateBillingGroupCommand: CreateBillingGroupCommand_1.CreateBillingGroupCommand,
    CreateCertificateFromCsrCommand: CreateCertificateFromCsrCommand_1.CreateCertificateFromCsrCommand,
    CreateCustomMetricCommand: CreateCustomMetricCommand_1.CreateCustomMetricCommand,
    CreateDimensionCommand: CreateDimensionCommand_1.CreateDimensionCommand,
    CreateDomainConfigurationCommand: CreateDomainConfigurationCommand_1.CreateDomainConfigurationCommand,
    CreateDynamicThingGroupCommand: CreateDynamicThingGroupCommand_1.CreateDynamicThingGroupCommand,
    CreateFleetMetricCommand: CreateFleetMetricCommand_1.CreateFleetMetricCommand,
    CreateJobCommand: CreateJobCommand_1.CreateJobCommand,
    CreateJobTemplateCommand: CreateJobTemplateCommand_1.CreateJobTemplateCommand,
    CreateKeysAndCertificateCommand: CreateKeysAndCertificateCommand_1.CreateKeysAndCertificateCommand,
    CreateMitigationActionCommand: CreateMitigationActionCommand_1.CreateMitigationActionCommand,
    CreateOTAUpdateCommand: CreateOTAUpdateCommand_1.CreateOTAUpdateCommand,
    CreatePackageCommand: CreatePackageCommand_1.CreatePackageCommand,
    CreatePackageVersionCommand: CreatePackageVersionCommand_1.CreatePackageVersionCommand,
    CreatePolicyCommand: CreatePolicyCommand_1.CreatePolicyCommand,
    CreatePolicyVersionCommand: CreatePolicyVersionCommand_1.CreatePolicyVersionCommand,
    CreateProvisioningClaimCommand: CreateProvisioningClaimCommand_1.CreateProvisioningClaimCommand,
    CreateProvisioningTemplateCommand: CreateProvisioningTemplateCommand_1.CreateProvisioningTemplateCommand,
    CreateProvisioningTemplateVersionCommand: CreateProvisioningTemplateVersionCommand_1.CreateProvisioningTemplateVersionCommand,
    CreateRoleAliasCommand: CreateRoleAliasCommand_1.CreateRoleAliasCommand,
    CreateScheduledAuditCommand: CreateScheduledAuditCommand_1.CreateScheduledAuditCommand,
    CreateSecurityProfileCommand: CreateSecurityProfileCommand_1.CreateSecurityProfileCommand,
    CreateStreamCommand: CreateStreamCommand_1.CreateStreamCommand,
    CreateThingCommand: CreateThingCommand_1.CreateThingCommand,
    CreateThingGroupCommand: CreateThingGroupCommand_1.CreateThingGroupCommand,
    CreateThingTypeCommand: CreateThingTypeCommand_1.CreateThingTypeCommand,
    CreateTopicRuleCommand: CreateTopicRuleCommand_1.CreateTopicRuleCommand,
    CreateTopicRuleDestinationCommand: CreateTopicRuleDestinationCommand_1.CreateTopicRuleDestinationCommand,
    DeleteAccountAuditConfigurationCommand: DeleteAccountAuditConfigurationCommand_1.DeleteAccountAuditConfigurationCommand,
    DeleteAuditSuppressionCommand: DeleteAuditSuppressionCommand_1.DeleteAuditSuppressionCommand,
    DeleteAuthorizerCommand: DeleteAuthorizerCommand_1.DeleteAuthorizerCommand,
    DeleteBillingGroupCommand: DeleteBillingGroupCommand_1.DeleteBillingGroupCommand,
    DeleteCACertificateCommand: DeleteCACertificateCommand_1.DeleteCACertificateCommand,
    DeleteCertificateCommand: DeleteCertificateCommand_1.DeleteCertificateCommand,
    DeleteCustomMetricCommand: DeleteCustomMetricCommand_1.DeleteCustomMetricCommand,
    DeleteDimensionCommand: DeleteDimensionCommand_1.DeleteDimensionCommand,
    DeleteDomainConfigurationCommand: DeleteDomainConfigurationCommand_1.DeleteDomainConfigurationCommand,
    DeleteDynamicThingGroupCommand: DeleteDynamicThingGroupCommand_1.DeleteDynamicThingGroupCommand,
    DeleteFleetMetricCommand: DeleteFleetMetricCommand_1.DeleteFleetMetricCommand,
    DeleteJobCommand: DeleteJobCommand_1.DeleteJobCommand,
    DeleteJobExecutionCommand: DeleteJobExecutionCommand_1.DeleteJobExecutionCommand,
    DeleteJobTemplateCommand: DeleteJobTemplateCommand_1.DeleteJobTemplateCommand,
    DeleteMitigationActionCommand: DeleteMitigationActionCommand_1.DeleteMitigationActionCommand,
    DeleteOTAUpdateCommand: DeleteOTAUpdateCommand_1.DeleteOTAUpdateCommand,
    DeletePackageCommand: DeletePackageCommand_1.DeletePackageCommand,
    DeletePackageVersionCommand: DeletePackageVersionCommand_1.DeletePackageVersionCommand,
    DeletePolicyCommand: DeletePolicyCommand_1.DeletePolicyCommand,
    DeletePolicyVersionCommand: DeletePolicyVersionCommand_1.DeletePolicyVersionCommand,
    DeleteProvisioningTemplateCommand: DeleteProvisioningTemplateCommand_1.DeleteProvisioningTemplateCommand,
    DeleteProvisioningTemplateVersionCommand: DeleteProvisioningTemplateVersionCommand_1.DeleteProvisioningTemplateVersionCommand,
    DeleteRegistrationCodeCommand: DeleteRegistrationCodeCommand_1.DeleteRegistrationCodeCommand,
    DeleteRoleAliasCommand: DeleteRoleAliasCommand_1.DeleteRoleAliasCommand,
    DeleteScheduledAuditCommand: DeleteScheduledAuditCommand_1.DeleteScheduledAuditCommand,
    DeleteSecurityProfileCommand: DeleteSecurityProfileCommand_1.DeleteSecurityProfileCommand,
    DeleteStreamCommand: DeleteStreamCommand_1.DeleteStreamCommand,
    DeleteThingCommand: DeleteThingCommand_1.DeleteThingCommand,
    DeleteThingGroupCommand: DeleteThingGroupCommand_1.DeleteThingGroupCommand,
    DeleteThingTypeCommand: DeleteThingTypeCommand_1.DeleteThingTypeCommand,
    DeleteTopicRuleCommand: DeleteTopicRuleCommand_1.DeleteTopicRuleCommand,
    DeleteTopicRuleDestinationCommand: DeleteTopicRuleDestinationCommand_1.DeleteTopicRuleDestinationCommand,
    DeleteV2LoggingLevelCommand: DeleteV2LoggingLevelCommand_1.DeleteV2LoggingLevelCommand,
    DeprecateThingTypeCommand: DeprecateThingTypeCommand_1.DeprecateThingTypeCommand,
    DescribeAccountAuditConfigurationCommand: DescribeAccountAuditConfigurationCommand_1.DescribeAccountAuditConfigurationCommand,
    DescribeAuditFindingCommand: DescribeAuditFindingCommand_1.DescribeAuditFindingCommand,
    DescribeAuditMitigationActionsTaskCommand: DescribeAuditMitigationActionsTaskCommand_1.DescribeAuditMitigationActionsTaskCommand,
    DescribeAuditSuppressionCommand: DescribeAuditSuppressionCommand_1.DescribeAuditSuppressionCommand,
    DescribeAuditTaskCommand: DescribeAuditTaskCommand_1.DescribeAuditTaskCommand,
    DescribeAuthorizerCommand: DescribeAuthorizerCommand_1.DescribeAuthorizerCommand,
    DescribeBillingGroupCommand: DescribeBillingGroupCommand_1.DescribeBillingGroupCommand,
    DescribeCACertificateCommand: DescribeCACertificateCommand_1.DescribeCACertificateCommand,
    DescribeCertificateCommand: DescribeCertificateCommand_1.DescribeCertificateCommand,
    DescribeCustomMetricCommand: DescribeCustomMetricCommand_1.DescribeCustomMetricCommand,
    DescribeDefaultAuthorizerCommand: DescribeDefaultAuthorizerCommand_1.DescribeDefaultAuthorizerCommand,
    DescribeDetectMitigationActionsTaskCommand: DescribeDetectMitigationActionsTaskCommand_1.DescribeDetectMitigationActionsTaskCommand,
    DescribeDimensionCommand: DescribeDimensionCommand_1.DescribeDimensionCommand,
    DescribeDomainConfigurationCommand: DescribeDomainConfigurationCommand_1.DescribeDomainConfigurationCommand,
    DescribeEndpointCommand: DescribeEndpointCommand_1.DescribeEndpointCommand,
    DescribeEventConfigurationsCommand: DescribeEventConfigurationsCommand_1.DescribeEventConfigurationsCommand,
    DescribeFleetMetricCommand: DescribeFleetMetricCommand_1.DescribeFleetMetricCommand,
    DescribeIndexCommand: DescribeIndexCommand_1.DescribeIndexCommand,
    DescribeJobCommand: DescribeJobCommand_1.DescribeJobCommand,
    DescribeJobExecutionCommand: DescribeJobExecutionCommand_1.DescribeJobExecutionCommand,
    DescribeJobTemplateCommand: DescribeJobTemplateCommand_1.DescribeJobTemplateCommand,
    DescribeManagedJobTemplateCommand: DescribeManagedJobTemplateCommand_1.DescribeManagedJobTemplateCommand,
    DescribeMitigationActionCommand: DescribeMitigationActionCommand_1.DescribeMitigationActionCommand,
    DescribeProvisioningTemplateCommand: DescribeProvisioningTemplateCommand_1.DescribeProvisioningTemplateCommand,
    DescribeProvisioningTemplateVersionCommand: DescribeProvisioningTemplateVersionCommand_1.DescribeProvisioningTemplateVersionCommand,
    DescribeRoleAliasCommand: DescribeRoleAliasCommand_1.DescribeRoleAliasCommand,
    DescribeScheduledAuditCommand: DescribeScheduledAuditCommand_1.DescribeScheduledAuditCommand,
    DescribeSecurityProfileCommand: DescribeSecurityProfileCommand_1.DescribeSecurityProfileCommand,
    DescribeStreamCommand: DescribeStreamCommand_1.DescribeStreamCommand,
    DescribeThingCommand: DescribeThingCommand_1.DescribeThingCommand,
    DescribeThingGroupCommand: DescribeThingGroupCommand_1.DescribeThingGroupCommand,
    DescribeThingRegistrationTaskCommand: DescribeThingRegistrationTaskCommand_1.DescribeThingRegistrationTaskCommand,
    DescribeThingTypeCommand: DescribeThingTypeCommand_1.DescribeThingTypeCommand,
    DetachPolicyCommand: DetachPolicyCommand_1.DetachPolicyCommand,
    DetachPrincipalPolicyCommand: DetachPrincipalPolicyCommand_1.DetachPrincipalPolicyCommand,
    DetachSecurityProfileCommand: DetachSecurityProfileCommand_1.DetachSecurityProfileCommand,
    DetachThingPrincipalCommand: DetachThingPrincipalCommand_1.DetachThingPrincipalCommand,
    DisableTopicRuleCommand: DisableTopicRuleCommand_1.DisableTopicRuleCommand,
    EnableTopicRuleCommand: EnableTopicRuleCommand_1.EnableTopicRuleCommand,
    GetBehaviorModelTrainingSummariesCommand: GetBehaviorModelTrainingSummariesCommand_1.GetBehaviorModelTrainingSummariesCommand,
    GetBucketsAggregationCommand: GetBucketsAggregationCommand_1.GetBucketsAggregationCommand,
    GetCardinalityCommand: GetCardinalityCommand_1.GetCardinalityCommand,
    GetEffectivePoliciesCommand: GetEffectivePoliciesCommand_1.GetEffectivePoliciesCommand,
    GetIndexingConfigurationCommand: GetIndexingConfigurationCommand_1.GetIndexingConfigurationCommand,
    GetJobDocumentCommand: GetJobDocumentCommand_1.GetJobDocumentCommand,
    GetLoggingOptionsCommand: GetLoggingOptionsCommand_1.GetLoggingOptionsCommand,
    GetOTAUpdateCommand: GetOTAUpdateCommand_1.GetOTAUpdateCommand,
    GetPackageCommand: GetPackageCommand_1.GetPackageCommand,
    GetPackageConfigurationCommand: GetPackageConfigurationCommand_1.GetPackageConfigurationCommand,
    GetPackageVersionCommand: GetPackageVersionCommand_1.GetPackageVersionCommand,
    GetPercentilesCommand: GetPercentilesCommand_1.GetPercentilesCommand,
    GetPolicyCommand: GetPolicyCommand_1.GetPolicyCommand,
    GetPolicyVersionCommand: GetPolicyVersionCommand_1.GetPolicyVersionCommand,
    GetRegistrationCodeCommand: GetRegistrationCodeCommand_1.GetRegistrationCodeCommand,
    GetStatisticsCommand: GetStatisticsCommand_1.GetStatisticsCommand,
    GetTopicRuleCommand: GetTopicRuleCommand_1.GetTopicRuleCommand,
    GetTopicRuleDestinationCommand: GetTopicRuleDestinationCommand_1.GetTopicRuleDestinationCommand,
    GetV2LoggingOptionsCommand: GetV2LoggingOptionsCommand_1.GetV2LoggingOptionsCommand,
    ListActiveViolationsCommand: ListActiveViolationsCommand_1.ListActiveViolationsCommand,
    ListAttachedPoliciesCommand: ListAttachedPoliciesCommand_1.ListAttachedPoliciesCommand,
    ListAuditFindingsCommand: ListAuditFindingsCommand_1.ListAuditFindingsCommand,
    ListAuditMitigationActionsExecutionsCommand: ListAuditMitigationActionsExecutionsCommand_1.ListAuditMitigationActionsExecutionsCommand,
    ListAuditMitigationActionsTasksCommand: ListAuditMitigationActionsTasksCommand_1.ListAuditMitigationActionsTasksCommand,
    ListAuditSuppressionsCommand: ListAuditSuppressionsCommand_1.ListAuditSuppressionsCommand,
    ListAuditTasksCommand: ListAuditTasksCommand_1.ListAuditTasksCommand,
    ListAuthorizersCommand: ListAuthorizersCommand_1.ListAuthorizersCommand,
    ListBillingGroupsCommand: ListBillingGroupsCommand_1.ListBillingGroupsCommand,
    ListCACertificatesCommand: ListCACertificatesCommand_1.ListCACertificatesCommand,
    ListCertificatesCommand: ListCertificatesCommand_1.ListCertificatesCommand,
    ListCertificatesByCACommand: ListCertificatesByCACommand_1.ListCertificatesByCACommand,
    ListCustomMetricsCommand: ListCustomMetricsCommand_1.ListCustomMetricsCommand,
    ListDetectMitigationActionsExecutionsCommand: ListDetectMitigationActionsExecutionsCommand_1.ListDetectMitigationActionsExecutionsCommand,
    ListDetectMitigationActionsTasksCommand: ListDetectMitigationActionsTasksCommand_1.ListDetectMitigationActionsTasksCommand,
    ListDimensionsCommand: ListDimensionsCommand_1.ListDimensionsCommand,
    ListDomainConfigurationsCommand: ListDomainConfigurationsCommand_1.ListDomainConfigurationsCommand,
    ListFleetMetricsCommand: ListFleetMetricsCommand_1.ListFleetMetricsCommand,
    ListIndicesCommand: ListIndicesCommand_1.ListIndicesCommand,
    ListJobExecutionsForJobCommand: ListJobExecutionsForJobCommand_1.ListJobExecutionsForJobCommand,
    ListJobExecutionsForThingCommand: ListJobExecutionsForThingCommand_1.ListJobExecutionsForThingCommand,
    ListJobsCommand: ListJobsCommand_1.ListJobsCommand,
    ListJobTemplatesCommand: ListJobTemplatesCommand_1.ListJobTemplatesCommand,
    ListManagedJobTemplatesCommand: ListManagedJobTemplatesCommand_1.ListManagedJobTemplatesCommand,
    ListMetricValuesCommand: ListMetricValuesCommand_1.ListMetricValuesCommand,
    ListMitigationActionsCommand: ListMitigationActionsCommand_1.ListMitigationActionsCommand,
    ListOTAUpdatesCommand: ListOTAUpdatesCommand_1.ListOTAUpdatesCommand,
    ListOutgoingCertificatesCommand: ListOutgoingCertificatesCommand_1.ListOutgoingCertificatesCommand,
    ListPackagesCommand: ListPackagesCommand_1.ListPackagesCommand,
    ListPackageVersionsCommand: ListPackageVersionsCommand_1.ListPackageVersionsCommand,
    ListPoliciesCommand: ListPoliciesCommand_1.ListPoliciesCommand,
    ListPolicyPrincipalsCommand: ListPolicyPrincipalsCommand_1.ListPolicyPrincipalsCommand,
    ListPolicyVersionsCommand: ListPolicyVersionsCommand_1.ListPolicyVersionsCommand,
    ListPrincipalPoliciesCommand: ListPrincipalPoliciesCommand_1.ListPrincipalPoliciesCommand,
    ListPrincipalThingsCommand: ListPrincipalThingsCommand_1.ListPrincipalThingsCommand,
    ListProvisioningTemplatesCommand: ListProvisioningTemplatesCommand_1.ListProvisioningTemplatesCommand,
    ListProvisioningTemplateVersionsCommand: ListProvisioningTemplateVersionsCommand_1.ListProvisioningTemplateVersionsCommand,
    ListRelatedResourcesForAuditFindingCommand: ListRelatedResourcesForAuditFindingCommand_1.ListRelatedResourcesForAuditFindingCommand,
    ListRoleAliasesCommand: ListRoleAliasesCommand_1.ListRoleAliasesCommand,
    ListScheduledAuditsCommand: ListScheduledAuditsCommand_1.ListScheduledAuditsCommand,
    ListSecurityProfilesCommand: ListSecurityProfilesCommand_1.ListSecurityProfilesCommand,
    ListSecurityProfilesForTargetCommand: ListSecurityProfilesForTargetCommand_1.ListSecurityProfilesForTargetCommand,
    ListStreamsCommand: ListStreamsCommand_1.ListStreamsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    ListTargetsForPolicyCommand: ListTargetsForPolicyCommand_1.ListTargetsForPolicyCommand,
    ListTargetsForSecurityProfileCommand: ListTargetsForSecurityProfileCommand_1.ListTargetsForSecurityProfileCommand,
    ListThingGroupsCommand: ListThingGroupsCommand_1.ListThingGroupsCommand,
    ListThingGroupsForThingCommand: ListThingGroupsForThingCommand_1.ListThingGroupsForThingCommand,
    ListThingPrincipalsCommand: ListThingPrincipalsCommand_1.ListThingPrincipalsCommand,
    ListThingRegistrationTaskReportsCommand: ListThingRegistrationTaskReportsCommand_1.ListThingRegistrationTaskReportsCommand,
    ListThingRegistrationTasksCommand: ListThingRegistrationTasksCommand_1.ListThingRegistrationTasksCommand,
    ListThingsCommand: ListThingsCommand_1.ListThingsCommand,
    ListThingsInBillingGroupCommand: ListThingsInBillingGroupCommand_1.ListThingsInBillingGroupCommand,
    ListThingsInThingGroupCommand: ListThingsInThingGroupCommand_1.ListThingsInThingGroupCommand,
    ListThingTypesCommand: ListThingTypesCommand_1.ListThingTypesCommand,
    ListTopicRuleDestinationsCommand: ListTopicRuleDestinationsCommand_1.ListTopicRuleDestinationsCommand,
    ListTopicRulesCommand: ListTopicRulesCommand_1.ListTopicRulesCommand,
    ListV2LoggingLevelsCommand: ListV2LoggingLevelsCommand_1.ListV2LoggingLevelsCommand,
    ListViolationEventsCommand: ListViolationEventsCommand_1.ListViolationEventsCommand,
    PutVerificationStateOnViolationCommand: PutVerificationStateOnViolationCommand_1.PutVerificationStateOnViolationCommand,
    RegisterCACertificateCommand: RegisterCACertificateCommand_1.RegisterCACertificateCommand,
    RegisterCertificateCommand: RegisterCertificateCommand_1.RegisterCertificateCommand,
    RegisterCertificateWithoutCACommand: RegisterCertificateWithoutCACommand_1.RegisterCertificateWithoutCACommand,
    RegisterThingCommand: RegisterThingCommand_1.RegisterThingCommand,
    RejectCertificateTransferCommand: RejectCertificateTransferCommand_1.RejectCertificateTransferCommand,
    RemoveThingFromBillingGroupCommand: RemoveThingFromBillingGroupCommand_1.RemoveThingFromBillingGroupCommand,
    RemoveThingFromThingGroupCommand: RemoveThingFromThingGroupCommand_1.RemoveThingFromThingGroupCommand,
    ReplaceTopicRuleCommand: ReplaceTopicRuleCommand_1.ReplaceTopicRuleCommand,
    SearchIndexCommand: SearchIndexCommand_1.SearchIndexCommand,
    SetDefaultAuthorizerCommand: SetDefaultAuthorizerCommand_1.SetDefaultAuthorizerCommand,
    SetDefaultPolicyVersionCommand: SetDefaultPolicyVersionCommand_1.SetDefaultPolicyVersionCommand,
    SetLoggingOptionsCommand: SetLoggingOptionsCommand_1.SetLoggingOptionsCommand,
    SetV2LoggingLevelCommand: SetV2LoggingLevelCommand_1.SetV2LoggingLevelCommand,
    SetV2LoggingOptionsCommand: SetV2LoggingOptionsCommand_1.SetV2LoggingOptionsCommand,
    StartAuditMitigationActionsTaskCommand: StartAuditMitigationActionsTaskCommand_1.StartAuditMitigationActionsTaskCommand,
    StartDetectMitigationActionsTaskCommand: StartDetectMitigationActionsTaskCommand_1.StartDetectMitigationActionsTaskCommand,
    StartOnDemandAuditTaskCommand: StartOnDemandAuditTaskCommand_1.StartOnDemandAuditTaskCommand,
    StartThingRegistrationTaskCommand: StartThingRegistrationTaskCommand_1.StartThingRegistrationTaskCommand,
    StopThingRegistrationTaskCommand: StopThingRegistrationTaskCommand_1.StopThingRegistrationTaskCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    TestAuthorizationCommand: TestAuthorizationCommand_1.TestAuthorizationCommand,
    TestInvokeAuthorizerCommand: TestInvokeAuthorizerCommand_1.TestInvokeAuthorizerCommand,
    TransferCertificateCommand: TransferCertificateCommand_1.TransferCertificateCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateAccountAuditConfigurationCommand: UpdateAccountAuditConfigurationCommand_1.UpdateAccountAuditConfigurationCommand,
    UpdateAuditSuppressionCommand: UpdateAuditSuppressionCommand_1.UpdateAuditSuppressionCommand,
    UpdateAuthorizerCommand: UpdateAuthorizerCommand_1.UpdateAuthorizerCommand,
    UpdateBillingGroupCommand: UpdateBillingGroupCommand_1.UpdateBillingGroupCommand,
    UpdateCACertificateCommand: UpdateCACertificateCommand_1.UpdateCACertificateCommand,
    UpdateCertificateCommand: UpdateCertificateCommand_1.UpdateCertificateCommand,
    UpdateCustomMetricCommand: UpdateCustomMetricCommand_1.UpdateCustomMetricCommand,
    UpdateDimensionCommand: UpdateDimensionCommand_1.UpdateDimensionCommand,
    UpdateDomainConfigurationCommand: UpdateDomainConfigurationCommand_1.UpdateDomainConfigurationCommand,
    UpdateDynamicThingGroupCommand: UpdateDynamicThingGroupCommand_1.UpdateDynamicThingGroupCommand,
    UpdateEventConfigurationsCommand: UpdateEventConfigurationsCommand_1.UpdateEventConfigurationsCommand,
    UpdateFleetMetricCommand: UpdateFleetMetricCommand_1.UpdateFleetMetricCommand,
    UpdateIndexingConfigurationCommand: UpdateIndexingConfigurationCommand_1.UpdateIndexingConfigurationCommand,
    UpdateJobCommand: UpdateJobCommand_1.UpdateJobCommand,
    UpdateMitigationActionCommand: UpdateMitigationActionCommand_1.UpdateMitigationActionCommand,
    UpdatePackageCommand: UpdatePackageCommand_1.UpdatePackageCommand,
    UpdatePackageConfigurationCommand: UpdatePackageConfigurationCommand_1.UpdatePackageConfigurationCommand,
    UpdatePackageVersionCommand: UpdatePackageVersionCommand_1.UpdatePackageVersionCommand,
    UpdateProvisioningTemplateCommand: UpdateProvisioningTemplateCommand_1.UpdateProvisioningTemplateCommand,
    UpdateRoleAliasCommand: UpdateRoleAliasCommand_1.UpdateRoleAliasCommand,
    UpdateScheduledAuditCommand: UpdateScheduledAuditCommand_1.UpdateScheduledAuditCommand,
    UpdateSecurityProfileCommand: UpdateSecurityProfileCommand_1.UpdateSecurityProfileCommand,
    UpdateStreamCommand: UpdateStreamCommand_1.UpdateStreamCommand,
    UpdateThingCommand: UpdateThingCommand_1.UpdateThingCommand,
    UpdateThingGroupCommand: UpdateThingGroupCommand_1.UpdateThingGroupCommand,
    UpdateThingGroupsForThingCommand: UpdateThingGroupsForThingCommand_1.UpdateThingGroupsForThingCommand,
    UpdateTopicRuleDestinationCommand: UpdateTopicRuleDestinationCommand_1.UpdateTopicRuleDestinationCommand,
    ValidateSecurityProfileBehaviorsCommand: ValidateSecurityProfileBehaviorsCommand_1.ValidateSecurityProfileBehaviorsCommand,
};
class IoT extends IoTClient_1.IoTClient {
}
exports.IoT = IoT;
(0, smithy_client_1.createAggregatedClient)(commands, IoT);
