// kysely-codegen@0.10.1/dist/dialects/postgres/postgres-dialect.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgresDialect = void 0;
const kysely_1 = require("kysely");
const dialect_1 = require("../../dialect");
const postgres_adapter_1 = require("./postgres-adapter");
const postgres_introspector_1 = require("./postgres-introspector");
class PostgresDialect extends dialect_1.Dialect {
    constructor() {
        super(...arguments);
        this.adapter = new postgres_adapter_1.PostgresAdapter();
        this.introspector = new postgres_introspector_1.PostgresIntrospector(this.adapter);
    }
    async createKyselyDialect(options) {
        const { Pool } = await Promise.resolve().then(() => __importStar(require('pg')));
        return new kysely_1.PostgresDialect({
            pool: new Pool({
                connectionString: options.connectionString,
                ssl: options.ssl ? { rejectUnauthorized: false } : false,
            }),
        });
    }
}
exports.PostgresDialect = PostgresDialect;
