// @aws-sdk/client-iot-data-plane@3.449.0/dist-cjs/IoTDataPlane.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoTDataPlane = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const DeleteThingShadowCommand_1 = require("./commands/DeleteThingShadowCommand");
const GetRetainedMessageCommand_1 = require("./commands/GetRetainedMessageCommand");
const GetThingShadowCommand_1 = require("./commands/GetThingShadowCommand");
const ListNamedShadowsForThingCommand_1 = require("./commands/ListNamedShadowsForThingCommand");
const ListRetainedMessagesCommand_1 = require("./commands/ListRetainedMessagesCommand");
const PublishCommand_1 = require("./commands/PublishCommand");
const UpdateThingShadowCommand_1 = require("./commands/UpdateThingShadowCommand");
const IoTDataPlaneClient_1 = require("./IoTDataPlaneClient");
const commands = {
    DeleteThingShadowCommand: DeleteThingShadowCommand_1.DeleteThingShadowCommand,
    GetRetainedMessageCommand: GetRetainedMessageCommand_1.GetRetainedMessageCommand,
    GetThingShadowCommand: GetThingShadowCommand_1.GetThingShadowCommand,
    ListNamedShadowsForThingCommand: ListNamedShadowsForThingCommand_1.ListNamedShadowsForThingCommand,
    ListRetainedMessagesCommand: ListRetainedMessagesCommand_1.ListRetainedMessagesCommand,
    PublishCommand: PublishCommand_1.PublishCommand,
    UpdateThingShadowCommand: UpdateThingShadowCommand_1.UpdateThingShadowCommand,
};
class IoTDataPlane extends IoTDataPlaneClient_1.IoTDataPlaneClient {
}
exports.IoTDataPlane = IoTDataPlane;
(0, smithy_client_1.createAggregatedClient)(commands, IoTDataPlane);
