// kysely-codegen@0.10.1/dist/enums/node-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeType = void 0;
var NodeType;
(function (NodeType) {
    NodeType["ALIAS_DECLARATION"] = "AliasDeclaration";
    NodeType["ARRAY_EXPRESSION"] = "ArrayExpression";
    NodeType["EXPORT_STATEMENT"] = "ExportStatement";
    NodeType["EXTENDS_CLAUSE"] = "ExtendsClause";
    NodeType["GENERIC_EXPRESSION"] = "GenericExpression";
    NodeType["IDENTIFIER"] = "Identifier";
    NodeType["IMPORT_CLAUSE"] = "ImportClause";
    NodeType["IMPORT_STATEMENT"] = "ImportStatement";
    NodeType["INFER_CLAUSE"] = "InferClause";
    NodeType["INTERFACE_DECLARATION"] = "InterfaceDeclaration";
    NodeType["LITERAL"] = "Literal";
    NodeType["MAPPED_TYPE"] = "MappedType";
    NodeType["MODULE_REFERENCE"] = "ModuleReference";
    NodeType["OBJECT_EXPRESSION"] = "ObjectExpression";
    NodeType["PROPERTY"] = "Property";
    NodeType["TEMPLATE"] = "Template";
    NodeType["UNION_EXPRESSION"] = "UnionExpression";
    NodeType["VALUE"] = "Value";
})(NodeType = exports.NodeType || (exports.NodeType = {}));
