// @aws-sdk/client-iam@3.449.0/dist-cjs/IAM.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAM = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const AddClientIDToOpenIDConnectProviderCommand_1 = require("./commands/AddClientIDToOpenIDConnectProviderCommand");
const AddRoleToInstanceProfileCommand_1 = require("./commands/AddRoleToInstanceProfileCommand");
const AddUserToGroupCommand_1 = require("./commands/AddUserToGroupCommand");
const AttachGroupPolicyCommand_1 = require("./commands/AttachGroupPolicyCommand");
const AttachRolePolicyCommand_1 = require("./commands/AttachRolePolicyCommand");
const AttachUserPolicyCommand_1 = require("./commands/AttachUserPolicyCommand");
const ChangePasswordCommand_1 = require("./commands/ChangePasswordCommand");
const CreateAccessKeyCommand_1 = require("./commands/CreateAccessKeyCommand");
const CreateAccountAliasCommand_1 = require("./commands/CreateAccountAliasCommand");
const CreateGroupCommand_1 = require("./commands/CreateGroupCommand");
const CreateInstanceProfileCommand_1 = require("./commands/CreateInstanceProfileCommand");
const CreateLoginProfileCommand_1 = require("./commands/CreateLoginProfileCommand");
const CreateOpenIDConnectProviderCommand_1 = require("./commands/CreateOpenIDConnectProviderCommand");
const CreatePolicyCommand_1 = require("./commands/CreatePolicyCommand");
const CreatePolicyVersionCommand_1 = require("./commands/CreatePolicyVersionCommand");
const CreateRoleCommand_1 = require("./commands/CreateRoleCommand");
const CreateSAMLProviderCommand_1 = require("./commands/CreateSAMLProviderCommand");
const CreateServiceLinkedRoleCommand_1 = require("./commands/CreateServiceLinkedRoleCommand");
const CreateServiceSpecificCredentialCommand_1 = require("./commands/CreateServiceSpecificCredentialCommand");
const CreateUserCommand_1 = require("./commands/CreateUserCommand");
const CreateVirtualMFADeviceCommand_1 = require("./commands/CreateVirtualMFADeviceCommand");
const DeactivateMFADeviceCommand_1 = require("./commands/DeactivateMFADeviceCommand");
const DeleteAccessKeyCommand_1 = require("./commands/DeleteAccessKeyCommand");
const DeleteAccountAliasCommand_1 = require("./commands/DeleteAccountAliasCommand");
const DeleteAccountPasswordPolicyCommand_1 = require("./commands/DeleteAccountPasswordPolicyCommand");
const DeleteGroupCommand_1 = require("./commands/DeleteGroupCommand");
const DeleteGroupPolicyCommand_1 = require("./commands/DeleteGroupPolicyCommand");
const DeleteInstanceProfileCommand_1 = require("./commands/DeleteInstanceProfileCommand");
const DeleteLoginProfileCommand_1 = require("./commands/DeleteLoginProfileCommand");
const DeleteOpenIDConnectProviderCommand_1 = require("./commands/DeleteOpenIDConnectProviderCommand");
const DeletePolicyCommand_1 = require("./commands/DeletePolicyCommand");
const DeletePolicyVersionCommand_1 = require("./commands/DeletePolicyVersionCommand");
const DeleteRoleCommand_1 = require("./commands/DeleteRoleCommand");
const DeleteRolePermissionsBoundaryCommand_1 = require("./commands/DeleteRolePermissionsBoundaryCommand");
const DeleteRolePolicyCommand_1 = require("./commands/DeleteRolePolicyCommand");
const DeleteSAMLProviderCommand_1 = require("./commands/DeleteSAMLProviderCommand");
const DeleteServerCertificateCommand_1 = require("./commands/DeleteServerCertificateCommand");
const DeleteServiceLinkedRoleCommand_1 = require("./commands/DeleteServiceLinkedRoleCommand");
const DeleteServiceSpecificCredentialCommand_1 = require("./commands/DeleteServiceSpecificCredentialCommand");
const DeleteSigningCertificateCommand_1 = require("./commands/DeleteSigningCertificateCommand");
const DeleteSSHPublicKeyCommand_1 = require("./commands/DeleteSSHPublicKeyCommand");
const DeleteUserCommand_1 = require("./commands/DeleteUserCommand");
const DeleteUserPermissionsBoundaryCommand_1 = require("./commands/DeleteUserPermissionsBoundaryCommand");
const DeleteUserPolicyCommand_1 = require("./commands/DeleteUserPolicyCommand");
const DeleteVirtualMFADeviceCommand_1 = require("./commands/DeleteVirtualMFADeviceCommand");
const DetachGroupPolicyCommand_1 = require("./commands/DetachGroupPolicyCommand");
const DetachRolePolicyCommand_1 = require("./commands/DetachRolePolicyCommand");
const DetachUserPolicyCommand_1 = require("./commands/DetachUserPolicyCommand");
const EnableMFADeviceCommand_1 = require("./commands/EnableMFADeviceCommand");
const GenerateCredentialReportCommand_1 = require("./commands/GenerateCredentialReportCommand");
const GenerateOrganizationsAccessReportCommand_1 = require("./commands/GenerateOrganizationsAccessReportCommand");
const GenerateServiceLastAccessedDetailsCommand_1 = require("./commands/GenerateServiceLastAccessedDetailsCommand");
const GetAccessKeyLastUsedCommand_1 = require("./commands/GetAccessKeyLastUsedCommand");
const GetAccountAuthorizationDetailsCommand_1 = require("./commands/GetAccountAuthorizationDetailsCommand");
const GetAccountPasswordPolicyCommand_1 = require("./commands/GetAccountPasswordPolicyCommand");
const GetAccountSummaryCommand_1 = require("./commands/GetAccountSummaryCommand");
const GetContextKeysForCustomPolicyCommand_1 = require("./commands/GetContextKeysForCustomPolicyCommand");
const GetContextKeysForPrincipalPolicyCommand_1 = require("./commands/GetContextKeysForPrincipalPolicyCommand");
const GetCredentialReportCommand_1 = require("./commands/GetCredentialReportCommand");
const GetGroupCommand_1 = require("./commands/GetGroupCommand");
const GetGroupPolicyCommand_1 = require("./commands/GetGroupPolicyCommand");
const GetInstanceProfileCommand_1 = require("./commands/GetInstanceProfileCommand");
const GetLoginProfileCommand_1 = require("./commands/GetLoginProfileCommand");
const GetMFADeviceCommand_1 = require("./commands/GetMFADeviceCommand");
const GetOpenIDConnectProviderCommand_1 = require("./commands/GetOpenIDConnectProviderCommand");
const GetOrganizationsAccessReportCommand_1 = require("./commands/GetOrganizationsAccessReportCommand");
const GetPolicyCommand_1 = require("./commands/GetPolicyCommand");
const GetPolicyVersionCommand_1 = require("./commands/GetPolicyVersionCommand");
const GetRoleCommand_1 = require("./commands/GetRoleCommand");
const GetRolePolicyCommand_1 = require("./commands/GetRolePolicyCommand");
const GetSAMLProviderCommand_1 = require("./commands/GetSAMLProviderCommand");
const GetServerCertificateCommand_1 = require("./commands/GetServerCertificateCommand");
const GetServiceLastAccessedDetailsCommand_1 = require("./commands/GetServiceLastAccessedDetailsCommand");
const GetServiceLastAccessedDetailsWithEntitiesCommand_1 = require("./commands/GetServiceLastAccessedDetailsWithEntitiesCommand");
const GetServiceLinkedRoleDeletionStatusCommand_1 = require("./commands/GetServiceLinkedRoleDeletionStatusCommand");
const GetSSHPublicKeyCommand_1 = require("./commands/GetSSHPublicKeyCommand");
const GetUserCommand_1 = require("./commands/GetUserCommand");
const GetUserPolicyCommand_1 = require("./commands/GetUserPolicyCommand");
const ListAccessKeysCommand_1 = require("./commands/ListAccessKeysCommand");
const ListAccountAliasesCommand_1 = require("./commands/ListAccountAliasesCommand");
const ListAttachedGroupPoliciesCommand_1 = require("./commands/ListAttachedGroupPoliciesCommand");
const ListAttachedRolePoliciesCommand_1 = require("./commands/ListAttachedRolePoliciesCommand");
const ListAttachedUserPoliciesCommand_1 = require("./commands/ListAttachedUserPoliciesCommand");
const ListEntitiesForPolicyCommand_1 = require("./commands/ListEntitiesForPolicyCommand");
const ListGroupPoliciesCommand_1 = require("./commands/ListGroupPoliciesCommand");
const ListGroupsCommand_1 = require("./commands/ListGroupsCommand");
const ListGroupsForUserCommand_1 = require("./commands/ListGroupsForUserCommand");
const ListInstanceProfilesCommand_1 = require("./commands/ListInstanceProfilesCommand");
const ListInstanceProfilesForRoleCommand_1 = require("./commands/ListInstanceProfilesForRoleCommand");
const ListInstanceProfileTagsCommand_1 = require("./commands/ListInstanceProfileTagsCommand");
const ListMFADevicesCommand_1 = require("./commands/ListMFADevicesCommand");
const ListMFADeviceTagsCommand_1 = require("./commands/ListMFADeviceTagsCommand");
const ListOpenIDConnectProvidersCommand_1 = require("./commands/ListOpenIDConnectProvidersCommand");
const ListOpenIDConnectProviderTagsCommand_1 = require("./commands/ListOpenIDConnectProviderTagsCommand");
const ListPoliciesCommand_1 = require("./commands/ListPoliciesCommand");
const ListPoliciesGrantingServiceAccessCommand_1 = require("./commands/ListPoliciesGrantingServiceAccessCommand");
const ListPolicyTagsCommand_1 = require("./commands/ListPolicyTagsCommand");
const ListPolicyVersionsCommand_1 = require("./commands/ListPolicyVersionsCommand");
const ListRolePoliciesCommand_1 = require("./commands/ListRolePoliciesCommand");
const ListRolesCommand_1 = require("./commands/ListRolesCommand");
const ListRoleTagsCommand_1 = require("./commands/ListRoleTagsCommand");
const ListSAMLProvidersCommand_1 = require("./commands/ListSAMLProvidersCommand");
const ListSAMLProviderTagsCommand_1 = require("./commands/ListSAMLProviderTagsCommand");
const ListServerCertificatesCommand_1 = require("./commands/ListServerCertificatesCommand");
const ListServerCertificateTagsCommand_1 = require("./commands/ListServerCertificateTagsCommand");
const ListServiceSpecificCredentialsCommand_1 = require("./commands/ListServiceSpecificCredentialsCommand");
const ListSigningCertificatesCommand_1 = require("./commands/ListSigningCertificatesCommand");
const ListSSHPublicKeysCommand_1 = require("./commands/ListSSHPublicKeysCommand");
const ListUserPoliciesCommand_1 = require("./commands/ListUserPoliciesCommand");
const ListUsersCommand_1 = require("./commands/ListUsersCommand");
const ListUserTagsCommand_1 = require("./commands/ListUserTagsCommand");
const ListVirtualMFADevicesCommand_1 = require("./commands/ListVirtualMFADevicesCommand");
const PutGroupPolicyCommand_1 = require("./commands/PutGroupPolicyCommand");
const PutRolePermissionsBoundaryCommand_1 = require("./commands/PutRolePermissionsBoundaryCommand");
const PutRolePolicyCommand_1 = require("./commands/PutRolePolicyCommand");
const PutUserPermissionsBoundaryCommand_1 = require("./commands/PutUserPermissionsBoundaryCommand");
const PutUserPolicyCommand_1 = require("./commands/PutUserPolicyCommand");
const RemoveClientIDFromOpenIDConnectProviderCommand_1 = require("./commands/RemoveClientIDFromOpenIDConnectProviderCommand");
const RemoveRoleFromInstanceProfileCommand_1 = require("./commands/RemoveRoleFromInstanceProfileCommand");
const RemoveUserFromGroupCommand_1 = require("./commands/RemoveUserFromGroupCommand");
const ResetServiceSpecificCredentialCommand_1 = require("./commands/ResetServiceSpecificCredentialCommand");
const ResyncMFADeviceCommand_1 = require("./commands/ResyncMFADeviceCommand");
const SetDefaultPolicyVersionCommand_1 = require("./commands/SetDefaultPolicyVersionCommand");
const SetSecurityTokenServicePreferencesCommand_1 = require("./commands/SetSecurityTokenServicePreferencesCommand");
const SimulateCustomPolicyCommand_1 = require("./commands/SimulateCustomPolicyCommand");
const SimulatePrincipalPolicyCommand_1 = require("./commands/SimulatePrincipalPolicyCommand");
const TagInstanceProfileCommand_1 = require("./commands/TagInstanceProfileCommand");
const TagMFADeviceCommand_1 = require("./commands/TagMFADeviceCommand");
const TagOpenIDConnectProviderCommand_1 = require("./commands/TagOpenIDConnectProviderCommand");
const TagPolicyCommand_1 = require("./commands/TagPolicyCommand");
const TagRoleCommand_1 = require("./commands/TagRoleCommand");
const TagSAMLProviderCommand_1 = require("./commands/TagSAMLProviderCommand");
const TagServerCertificateCommand_1 = require("./commands/TagServerCertificateCommand");
const TagUserCommand_1 = require("./commands/TagUserCommand");
const UntagInstanceProfileCommand_1 = require("./commands/UntagInstanceProfileCommand");
const UntagMFADeviceCommand_1 = require("./commands/UntagMFADeviceCommand");
const UntagOpenIDConnectProviderCommand_1 = require("./commands/UntagOpenIDConnectProviderCommand");
const UntagPolicyCommand_1 = require("./commands/UntagPolicyCommand");
const UntagRoleCommand_1 = require("./commands/UntagRoleCommand");
const UntagSAMLProviderCommand_1 = require("./commands/UntagSAMLProviderCommand");
const UntagServerCertificateCommand_1 = require("./commands/UntagServerCertificateCommand");
const UntagUserCommand_1 = require("./commands/UntagUserCommand");
const UpdateAccessKeyCommand_1 = require("./commands/UpdateAccessKeyCommand");
const UpdateAccountPasswordPolicyCommand_1 = require("./commands/UpdateAccountPasswordPolicyCommand");
const UpdateAssumeRolePolicyCommand_1 = require("./commands/UpdateAssumeRolePolicyCommand");
const UpdateGroupCommand_1 = require("./commands/UpdateGroupCommand");
const UpdateLoginProfileCommand_1 = require("./commands/UpdateLoginProfileCommand");
const UpdateOpenIDConnectProviderThumbprintCommand_1 = require("./commands/UpdateOpenIDConnectProviderThumbprintCommand");
const UpdateRoleCommand_1 = require("./commands/UpdateRoleCommand");
const UpdateRoleDescriptionCommand_1 = require("./commands/UpdateRoleDescriptionCommand");
const UpdateSAMLProviderCommand_1 = require("./commands/UpdateSAMLProviderCommand");
const UpdateServerCertificateCommand_1 = require("./commands/UpdateServerCertificateCommand");
const UpdateServiceSpecificCredentialCommand_1 = require("./commands/UpdateServiceSpecificCredentialCommand");
const UpdateSigningCertificateCommand_1 = require("./commands/UpdateSigningCertificateCommand");
const UpdateSSHPublicKeyCommand_1 = require("./commands/UpdateSSHPublicKeyCommand");
const UpdateUserCommand_1 = require("./commands/UpdateUserCommand");
const UploadServerCertificateCommand_1 = require("./commands/UploadServerCertificateCommand");
const UploadSigningCertificateCommand_1 = require("./commands/UploadSigningCertificateCommand");
const UploadSSHPublicKeyCommand_1 = require("./commands/UploadSSHPublicKeyCommand");
const IAMClient_1 = require("./IAMClient");
const commands = {
    AddClientIDToOpenIDConnectProviderCommand: AddClientIDToOpenIDConnectProviderCommand_1.AddClientIDToOpenIDConnectProviderCommand,
    AddRoleToInstanceProfileCommand: AddRoleToInstanceProfileCommand_1.AddRoleToInstanceProfileCommand,
    AddUserToGroupCommand: AddUserToGroupCommand_1.AddUserToGroupCommand,
    AttachGroupPolicyCommand: AttachGroupPolicyCommand_1.AttachGroupPolicyCommand,
    AttachRolePolicyCommand: AttachRolePolicyCommand_1.AttachRolePolicyCommand,
    AttachUserPolicyCommand: AttachUserPolicyCommand_1.AttachUserPolicyCommand,
    ChangePasswordCommand: ChangePasswordCommand_1.ChangePasswordCommand,
    CreateAccessKeyCommand: CreateAccessKeyCommand_1.CreateAccessKeyCommand,
    CreateAccountAliasCommand: CreateAccountAliasCommand_1.CreateAccountAliasCommand,
    CreateGroupCommand: CreateGroupCommand_1.CreateGroupCommand,
    CreateInstanceProfileCommand: CreateInstanceProfileCommand_1.CreateInstanceProfileCommand,
    CreateLoginProfileCommand: CreateLoginProfileCommand_1.CreateLoginProfileCommand,
    CreateOpenIDConnectProviderCommand: CreateOpenIDConnectProviderCommand_1.CreateOpenIDConnectProviderCommand,
    CreatePolicyCommand: CreatePolicyCommand_1.CreatePolicyCommand,
    CreatePolicyVersionCommand: CreatePolicyVersionCommand_1.CreatePolicyVersionCommand,
    CreateRoleCommand: CreateRoleCommand_1.CreateRoleCommand,
    CreateSAMLProviderCommand: CreateSAMLProviderCommand_1.CreateSAMLProviderCommand,
    CreateServiceLinkedRoleCommand: CreateServiceLinkedRoleCommand_1.CreateServiceLinkedRoleCommand,
    CreateServiceSpecificCredentialCommand: CreateServiceSpecificCredentialCommand_1.CreateServiceSpecificCredentialCommand,
    CreateUserCommand: CreateUserCommand_1.CreateUserCommand,
    CreateVirtualMFADeviceCommand: CreateVirtualMFADeviceCommand_1.CreateVirtualMFADeviceCommand,
    DeactivateMFADeviceCommand: DeactivateMFADeviceCommand_1.DeactivateMFADeviceCommand,
    DeleteAccessKeyCommand: DeleteAccessKeyCommand_1.DeleteAccessKeyCommand,
    DeleteAccountAliasCommand: DeleteAccountAliasCommand_1.DeleteAccountAliasCommand,
    DeleteAccountPasswordPolicyCommand: DeleteAccountPasswordPolicyCommand_1.DeleteAccountPasswordPolicyCommand,
    DeleteGroupCommand: DeleteGroupCommand_1.DeleteGroupCommand,
    DeleteGroupPolicyCommand: DeleteGroupPolicyCommand_1.DeleteGroupPolicyCommand,
    DeleteInstanceProfileCommand: DeleteInstanceProfileCommand_1.DeleteInstanceProfileCommand,
    DeleteLoginProfileCommand: DeleteLoginProfileCommand_1.DeleteLoginProfileCommand,
    DeleteOpenIDConnectProviderCommand: DeleteOpenIDConnectProviderCommand_1.DeleteOpenIDConnectProviderCommand,
    DeletePolicyCommand: DeletePolicyCommand_1.DeletePolicyCommand,
    DeletePolicyVersionCommand: DeletePolicyVersionCommand_1.DeletePolicyVersionCommand,
    DeleteRoleCommand: DeleteRoleCommand_1.DeleteRoleCommand,
    DeleteRolePermissionsBoundaryCommand: DeleteRolePermissionsBoundaryCommand_1.DeleteRolePermissionsBoundaryCommand,
    DeleteRolePolicyCommand: DeleteRolePolicyCommand_1.DeleteRolePolicyCommand,
    DeleteSAMLProviderCommand: DeleteSAMLProviderCommand_1.DeleteSAMLProviderCommand,
    DeleteServerCertificateCommand: DeleteServerCertificateCommand_1.DeleteServerCertificateCommand,
    DeleteServiceLinkedRoleCommand: DeleteServiceLinkedRoleCommand_1.DeleteServiceLinkedRoleCommand,
    DeleteServiceSpecificCredentialCommand: DeleteServiceSpecificCredentialCommand_1.DeleteServiceSpecificCredentialCommand,
    DeleteSigningCertificateCommand: DeleteSigningCertificateCommand_1.DeleteSigningCertificateCommand,
    DeleteSSHPublicKeyCommand: DeleteSSHPublicKeyCommand_1.DeleteSSHPublicKeyCommand,
    DeleteUserCommand: DeleteUserCommand_1.DeleteUserCommand,
    DeleteUserPermissionsBoundaryCommand: DeleteUserPermissionsBoundaryCommand_1.DeleteUserPermissionsBoundaryCommand,
    DeleteUserPolicyCommand: DeleteUserPolicyCommand_1.DeleteUserPolicyCommand,
    DeleteVirtualMFADeviceCommand: DeleteVirtualMFADeviceCommand_1.DeleteVirtualMFADeviceCommand,
    DetachGroupPolicyCommand: DetachGroupPolicyCommand_1.DetachGroupPolicyCommand,
    DetachRolePolicyCommand: DetachRolePolicyCommand_1.DetachRolePolicyCommand,
    DetachUserPolicyCommand: DetachUserPolicyCommand_1.DetachUserPolicyCommand,
    EnableMFADeviceCommand: EnableMFADeviceCommand_1.EnableMFADeviceCommand,
    GenerateCredentialReportCommand: GenerateCredentialReportCommand_1.GenerateCredentialReportCommand,
    GenerateOrganizationsAccessReportCommand: GenerateOrganizationsAccessReportCommand_1.GenerateOrganizationsAccessReportCommand,
    GenerateServiceLastAccessedDetailsCommand: GenerateServiceLastAccessedDetailsCommand_1.GenerateServiceLastAccessedDetailsCommand,
    GetAccessKeyLastUsedCommand: GetAccessKeyLastUsedCommand_1.GetAccessKeyLastUsedCommand,
    GetAccountAuthorizationDetailsCommand: GetAccountAuthorizationDetailsCommand_1.GetAccountAuthorizationDetailsCommand,
    GetAccountPasswordPolicyCommand: GetAccountPasswordPolicyCommand_1.GetAccountPasswordPolicyCommand,
    GetAccountSummaryCommand: GetAccountSummaryCommand_1.GetAccountSummaryCommand,
    GetContextKeysForCustomPolicyCommand: GetContextKeysForCustomPolicyCommand_1.GetContextKeysForCustomPolicyCommand,
    GetContextKeysForPrincipalPolicyCommand: GetContextKeysForPrincipalPolicyCommand_1.GetContextKeysForPrincipalPolicyCommand,
    GetCredentialReportCommand: GetCredentialReportCommand_1.GetCredentialReportCommand,
    GetGroupCommand: GetGroupCommand_1.GetGroupCommand,
    GetGroupPolicyCommand: GetGroupPolicyCommand_1.GetGroupPolicyCommand,
    GetInstanceProfileCommand: GetInstanceProfileCommand_1.GetInstanceProfileCommand,
    GetLoginProfileCommand: GetLoginProfileCommand_1.GetLoginProfileCommand,
    GetMFADeviceCommand: GetMFADeviceCommand_1.GetMFADeviceCommand,
    GetOpenIDConnectProviderCommand: GetOpenIDConnectProviderCommand_1.GetOpenIDConnectProviderCommand,
    GetOrganizationsAccessReportCommand: GetOrganizationsAccessReportCommand_1.GetOrganizationsAccessReportCommand,
    GetPolicyCommand: GetPolicyCommand_1.GetPolicyCommand,
    GetPolicyVersionCommand: GetPolicyVersionCommand_1.GetPolicyVersionCommand,
    GetRoleCommand: GetRoleCommand_1.GetRoleCommand,
    GetRolePolicyCommand: GetRolePolicyCommand_1.GetRolePolicyCommand,
    GetSAMLProviderCommand: GetSAMLProviderCommand_1.GetSAMLProviderCommand,
    GetServerCertificateCommand: GetServerCertificateCommand_1.GetServerCertificateCommand,
    GetServiceLastAccessedDetailsCommand: GetServiceLastAccessedDetailsCommand_1.GetServiceLastAccessedDetailsCommand,
    GetServiceLastAccessedDetailsWithEntitiesCommand: GetServiceLastAccessedDetailsWithEntitiesCommand_1.GetServiceLastAccessedDetailsWithEntitiesCommand,
    GetServiceLinkedRoleDeletionStatusCommand: GetServiceLinkedRoleDeletionStatusCommand_1.GetServiceLinkedRoleDeletionStatusCommand,
    GetSSHPublicKeyCommand: GetSSHPublicKeyCommand_1.GetSSHPublicKeyCommand,
    GetUserCommand: GetUserCommand_1.GetUserCommand,
    GetUserPolicyCommand: GetUserPolicyCommand_1.GetUserPolicyCommand,
    ListAccessKeysCommand: ListAccessKeysCommand_1.ListAccessKeysCommand,
    ListAccountAliasesCommand: ListAccountAliasesCommand_1.ListAccountAliasesCommand,
    ListAttachedGroupPoliciesCommand: ListAttachedGroupPoliciesCommand_1.ListAttachedGroupPoliciesCommand,
    ListAttachedRolePoliciesCommand: ListAttachedRolePoliciesCommand_1.ListAttachedRolePoliciesCommand,
    ListAttachedUserPoliciesCommand: ListAttachedUserPoliciesCommand_1.ListAttachedUserPoliciesCommand,
    ListEntitiesForPolicyCommand: ListEntitiesForPolicyCommand_1.ListEntitiesForPolicyCommand,
    ListGroupPoliciesCommand: ListGroupPoliciesCommand_1.ListGroupPoliciesCommand,
    ListGroupsCommand: ListGroupsCommand_1.ListGroupsCommand,
    ListGroupsForUserCommand: ListGroupsForUserCommand_1.ListGroupsForUserCommand,
    ListInstanceProfilesCommand: ListInstanceProfilesCommand_1.ListInstanceProfilesCommand,
    ListInstanceProfilesForRoleCommand: ListInstanceProfilesForRoleCommand_1.ListInstanceProfilesForRoleCommand,
    ListInstanceProfileTagsCommand: ListInstanceProfileTagsCommand_1.ListInstanceProfileTagsCommand,
    ListMFADevicesCommand: ListMFADevicesCommand_1.ListMFADevicesCommand,
    ListMFADeviceTagsCommand: ListMFADeviceTagsCommand_1.ListMFADeviceTagsCommand,
    ListOpenIDConnectProvidersCommand: ListOpenIDConnectProvidersCommand_1.ListOpenIDConnectProvidersCommand,
    ListOpenIDConnectProviderTagsCommand: ListOpenIDConnectProviderTagsCommand_1.ListOpenIDConnectProviderTagsCommand,
    ListPoliciesCommand: ListPoliciesCommand_1.ListPoliciesCommand,
    ListPoliciesGrantingServiceAccessCommand: ListPoliciesGrantingServiceAccessCommand_1.ListPoliciesGrantingServiceAccessCommand,
    ListPolicyTagsCommand: ListPolicyTagsCommand_1.ListPolicyTagsCommand,
    ListPolicyVersionsCommand: ListPolicyVersionsCommand_1.ListPolicyVersionsCommand,
    ListRolePoliciesCommand: ListRolePoliciesCommand_1.ListRolePoliciesCommand,
    ListRolesCommand: ListRolesCommand_1.ListRolesCommand,
    ListRoleTagsCommand: ListRoleTagsCommand_1.ListRoleTagsCommand,
    ListSAMLProvidersCommand: ListSAMLProvidersCommand_1.ListSAMLProvidersCommand,
    ListSAMLProviderTagsCommand: ListSAMLProviderTagsCommand_1.ListSAMLProviderTagsCommand,
    ListServerCertificatesCommand: ListServerCertificatesCommand_1.ListServerCertificatesCommand,
    ListServerCertificateTagsCommand: ListServerCertificateTagsCommand_1.ListServerCertificateTagsCommand,
    ListServiceSpecificCredentialsCommand: ListServiceSpecificCredentialsCommand_1.ListServiceSpecificCredentialsCommand,
    ListSigningCertificatesCommand: ListSigningCertificatesCommand_1.ListSigningCertificatesCommand,
    ListSSHPublicKeysCommand: ListSSHPublicKeysCommand_1.ListSSHPublicKeysCommand,
    ListUserPoliciesCommand: ListUserPoliciesCommand_1.ListUserPoliciesCommand,
    ListUsersCommand: ListUsersCommand_1.ListUsersCommand,
    ListUserTagsCommand: ListUserTagsCommand_1.ListUserTagsCommand,
    ListVirtualMFADevicesCommand: ListVirtualMFADevicesCommand_1.ListVirtualMFADevicesCommand,
    PutGroupPolicyCommand: PutGroupPolicyCommand_1.PutGroupPolicyCommand,
    PutRolePermissionsBoundaryCommand: PutRolePermissionsBoundaryCommand_1.PutRolePermissionsBoundaryCommand,
    PutRolePolicyCommand: PutRolePolicyCommand_1.PutRolePolicyCommand,
    PutUserPermissionsBoundaryCommand: PutUserPermissionsBoundaryCommand_1.PutUserPermissionsBoundaryCommand,
    PutUserPolicyCommand: PutUserPolicyCommand_1.PutUserPolicyCommand,
    RemoveClientIDFromOpenIDConnectProviderCommand: RemoveClientIDFromOpenIDConnectProviderCommand_1.RemoveClientIDFromOpenIDConnectProviderCommand,
    RemoveRoleFromInstanceProfileCommand: RemoveRoleFromInstanceProfileCommand_1.RemoveRoleFromInstanceProfileCommand,
    RemoveUserFromGroupCommand: RemoveUserFromGroupCommand_1.RemoveUserFromGroupCommand,
    ResetServiceSpecificCredentialCommand: ResetServiceSpecificCredentialCommand_1.ResetServiceSpecificCredentialCommand,
    ResyncMFADeviceCommand: ResyncMFADeviceCommand_1.ResyncMFADeviceCommand,
    SetDefaultPolicyVersionCommand: SetDefaultPolicyVersionCommand_1.SetDefaultPolicyVersionCommand,
    SetSecurityTokenServicePreferencesCommand: SetSecurityTokenServicePreferencesCommand_1.SetSecurityTokenServicePreferencesCommand,
    SimulateCustomPolicyCommand: SimulateCustomPolicyCommand_1.SimulateCustomPolicyCommand,
    SimulatePrincipalPolicyCommand: SimulatePrincipalPolicyCommand_1.SimulatePrincipalPolicyCommand,
    TagInstanceProfileCommand: TagInstanceProfileCommand_1.TagInstanceProfileCommand,
    TagMFADeviceCommand: TagMFADeviceCommand_1.TagMFADeviceCommand,
    TagOpenIDConnectProviderCommand: TagOpenIDConnectProviderCommand_1.TagOpenIDConnectProviderCommand,
    TagPolicyCommand: TagPolicyCommand_1.TagPolicyCommand,
    TagRoleCommand: TagRoleCommand_1.TagRoleCommand,
    TagSAMLProviderCommand: TagSAMLProviderCommand_1.TagSAMLProviderCommand,
    TagServerCertificateCommand: TagServerCertificateCommand_1.TagServerCertificateCommand,
    TagUserCommand: TagUserCommand_1.TagUserCommand,
    UntagInstanceProfileCommand: UntagInstanceProfileCommand_1.UntagInstanceProfileCommand,
    UntagMFADeviceCommand: UntagMFADeviceCommand_1.UntagMFADeviceCommand,
    UntagOpenIDConnectProviderCommand: UntagOpenIDConnectProviderCommand_1.UntagOpenIDConnectProviderCommand,
    UntagPolicyCommand: UntagPolicyCommand_1.UntagPolicyCommand,
    UntagRoleCommand: UntagRoleCommand_1.UntagRoleCommand,
    UntagSAMLProviderCommand: UntagSAMLProviderCommand_1.UntagSAMLProviderCommand,
    UntagServerCertificateCommand: UntagServerCertificateCommand_1.UntagServerCertificateCommand,
    UntagUserCommand: UntagUserCommand_1.UntagUserCommand,
    UpdateAccessKeyCommand: UpdateAccessKeyCommand_1.UpdateAccessKeyCommand,
    UpdateAccountPasswordPolicyCommand: UpdateAccountPasswordPolicyCommand_1.UpdateAccountPasswordPolicyCommand,
    UpdateAssumeRolePolicyCommand: UpdateAssumeRolePolicyCommand_1.UpdateAssumeRolePolicyCommand,
    UpdateGroupCommand: UpdateGroupCommand_1.UpdateGroupCommand,
    UpdateLoginProfileCommand: UpdateLoginProfileCommand_1.UpdateLoginProfileCommand,
    UpdateOpenIDConnectProviderThumbprintCommand: UpdateOpenIDConnectProviderThumbprintCommand_1.UpdateOpenIDConnectProviderThumbprintCommand,
    UpdateRoleCommand: UpdateRoleCommand_1.UpdateRoleCommand,
    UpdateRoleDescriptionCommand: UpdateRoleDescriptionCommand_1.UpdateRoleDescriptionCommand,
    UpdateSAMLProviderCommand: UpdateSAMLProviderCommand_1.UpdateSAMLProviderCommand,
    UpdateServerCertificateCommand: UpdateServerCertificateCommand_1.UpdateServerCertificateCommand,
    UpdateServiceSpecificCredentialCommand: UpdateServiceSpecificCredentialCommand_1.UpdateServiceSpecificCredentialCommand,
    UpdateSigningCertificateCommand: UpdateSigningCertificateCommand_1.UpdateSigningCertificateCommand,
    UpdateSSHPublicKeyCommand: UpdateSSHPublicKeyCommand_1.UpdateSSHPublicKeyCommand,
    UpdateUserCommand: UpdateUserCommand_1.UpdateUserCommand,
    UploadServerCertificateCommand: UploadServerCertificateCommand_1.UploadServerCertificateCommand,
    UploadSigningCertificateCommand: UploadSigningCertificateCommand_1.UploadSigningCertificateCommand,
    UploadSSHPublicKeyCommand: UploadSSHPublicKeyCommand_1.UploadSSHPublicKeyCommand,
};
class IAM extends IAMClient_1.IAMClient {
}
exports.IAM = IAM;
(0, smithy_client_1.createAggregatedClient)(commands, IAM);
