// cdk-assets@2.91.0/bin/cdk-assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const yargs = require("yargs");
const list_1 = require("./list");
const logging_1 = require("./logging");
const publish_1 = require("./publish");
const lib_1 = require("../lib");
async function main() {
    const argv = yargs
        .usage('$0 <cmd> [args]')
        .option('verbose', {
        alias: 'v',
        type: 'boolean',
        desc: 'Increase logging verbosity',
        count: true,
        default: 0,
    })
        .option('path', {
        alias: 'p',
        type: 'string',
        desc: 'The path (file or directory) to load the assets from. If a directory, ' +
            `the file '${lib_1.AssetManifest.DEFAULT_FILENAME}' will be loaded from it.`,
        default: '.',
        requiresArg: true,
    })
        .command('ls', 'List assets from the given manifest', command => command, wrapHandler(async (args) => {
        await (0, list_1.list)(args);
    }))
        .command('publish [ASSET..]', 'Publish assets in the given manifest', command => command
        .option('profile', { type: 'string', describe: 'Profile to use from AWS Credentials file' })
        .positional('ASSET', { type: 'string', array: true, describe: 'Assets to publish (format: "ASSET[:DEST]"), default all' }), wrapHandler(async (args) => {
        await (0, publish_1.publish)({
            path: args.path,
            assets: args.ASSET,
            profile: args.profile,
        });
    }))
        .demandCommand()
        .help()
        .strict() // Error on wrong command
        .version(logging_1.VERSION)
        .showHelpOnFail(false)
        .argv;
    // Evaluating .argv triggers the parsing but the command gets implicitly executed,
    // so we don't need the output.
    Array.isArray(argv);
}
/**
 * Wrap a command's handler with standard pre- and post-work
 */
function wrapHandler(handler) {
    return async (argv) => {
        if (argv.verbose) {
            (0, logging_1.setLogThreshold)('verbose');
        }
        await handler(argv);
    };
}
main().catch(e => {
    // eslint-disable-next-line no-console
    console.error(e.stack);
    process.exitCode = 1;
});
