// sst@2.36.1/constructs/util/functionBinding.js
import * as ssm from "aws-cdk-lib/aws-ssm";
import { Config } from "../../config.js";
export function bindEnvironment(c) {
    const binding = c.getFunctionBinding();
    let environment = {};
    if (binding) {
        Object.entries(binding.variables).forEach(([prop, variable]) => {
            const envName = getEnvironmentKey(c, prop);
            if (variable.type === "plain") {
                environment[envName] = variable.value;
            }
            else if (variable.type === "secret" || variable.type === "site_url") {
                environment[envName] = placeholderSecretValue();
            }
            else if (variable.type === "secret_reference") {
                environment[envName] = placeholderSecretReferenceValue(variable.secret);
            }
        });
    }
    return environment;
}
export function bindParameters(c) {
    const binding = c.getFunctionBinding();
    if (!binding) {
        return;
    }
    const app = c.node.root;
    Object.entries(binding.variables).forEach(([prop, variable]) => {
        const resId = `Parameter_${prop}`;
        if (!c.node.tryFindChild(resId)) {
            if (variable.type === "plain" || variable.type === "site_url") {
                new ssm.StringParameter(c, resId, {
                    parameterName: getParameterPath(c, prop),
                    stringValue: variable.value,
                });
            }
            else if (variable.type === "secret_reference") {
                new ssm.StringParameter(c, resId, {
                    parameterName: getParameterPath(c, prop),
                    stringValue: placeholderSecretReferenceValue(variable.secret),
                });
            }
        }
    });
}
export function bindPermissions(c) {
    const binding = c.getFunctionBinding();
    if (!binding) {
        return {};
    }
    return c.getFunctionBinding()?.permissions || {};
}
export function bindType(c) {
    const binding = c.getFunctionBinding();
    if (!binding) {
        return;
    }
    return {
        clientPackage: binding.clientPackage,
        variables: Object.keys(binding.variables),
    };
}
export function getReferencedSecrets(c) {
    const binding = c.getFunctionBinding();
    const secrets = [];
    if (binding) {
        Object.values(binding.variables).forEach((variable) => {
            if (variable.type === "secret_reference") {
                secrets.push(variable.secret);
            }
        });
    }
    return secrets;
}
export function getEnvironmentKey(c, prop) {
    return Config.envFor({
        type: c.constructor.name,
        id: c.id,
        prop: prop,
    });
}
export function getParameterPath(c, prop) {
    const construct = c.constructor.name;
    return Config.pathFor({
        id: c.id,
        type: construct,
        prop: prop,
    });
}
export function getParameterFallbackPath(c, prop) {
    const construct = c.constructor.name;
    return Config.pathFor({
        id: c.id,
        type: construct,
        prop: prop,
        fallback: true,
    });
}
export function placeholderSecretValue() {
    return "__FETCH_FROM_SSM__";
}
export function placeholderSecretReferenceValue(secret) {
    return "__FETCH_FROM_SECRET__:" + secret.name;
}
