// kysely-codegen@0.10.1/dist/constants/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSON_VALUE_DEFINITION = exports.JSON_PRIMITIVE_DEFINITION = exports.JSON_OBJECT_DEFINITION = exports.JSON_ARRAY_DEFINITION = exports.GLOBAL_DEFINITIONS = void 0;
const nodes_1 = require("../nodes");
exports.GLOBAL_DEFINITIONS = {
    Generated: new nodes_1.TemplateNode(['T'], new nodes_1.ExtendsClauseNode('T', new nodes_1.GenericExpressionNode('ColumnType', [
        new nodes_1.InferClauseNode('S'),
        new nodes_1.InferClauseNode('I'),
        new nodes_1.InferClauseNode('U'),
    ]), new nodes_1.GenericExpressionNode('ColumnType', [
        new nodes_1.IdentifierNode('S'),
        new nodes_1.UnionExpressionNode([
            new nodes_1.IdentifierNode('I'),
            new nodes_1.IdentifierNode('undefined'),
        ]),
        new nodes_1.IdentifierNode('U'),
    ]), new nodes_1.GenericExpressionNode('ColumnType', [
        new nodes_1.IdentifierNode('T'),
        new nodes_1.UnionExpressionNode([
            new nodes_1.IdentifierNode('T'),
            new nodes_1.IdentifierNode('undefined'),
        ]),
        new nodes_1.IdentifierNode('T'),
    ]))),
};
exports.JSON_ARRAY_DEFINITION = new nodes_1.ArrayExpressionNode(new nodes_1.IdentifierNode('JsonValue'));
exports.JSON_OBJECT_DEFINITION = new nodes_1.MappedTypeNode(new nodes_1.IdentifierNode('JsonValue'));
exports.JSON_PRIMITIVE_DEFINITION = new nodes_1.UnionExpressionNode([
    new nodes_1.IdentifierNode('boolean'),
    new nodes_1.IdentifierNode('null'),
    new nodes_1.IdentifierNode('number'),
    new nodes_1.IdentifierNode('string'),
]);
exports.JSON_VALUE_DEFINITION = new nodes_1.UnionExpressionNode([
    new nodes_1.IdentifierNode('JsonArray'),
    new nodes_1.IdentifierNode('JsonObject'),
    new nodes_1.IdentifierNode('JsonPrimitive'),
]);
